<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>An integrative multi-assay class for experiment data — MultiAssayExperiment-class • MultiAssayExperiment</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="An integrative multi-assay class for experiment data — MultiAssayExperiment-class" />
<meta property="og:description" content="The MultiAssayExperiment class can be used to manage results of
diverse assays on a collection of specimen. Currently,  the class can handle
assays that are organized instances of
SummarizedExperiment,
ExpressionSet, matrix,
RaggedExperiment
(inherits from GRangesList), and RangedVcfStack.
Create new MultiAssayExperiment instances with the homonymous
constructor, minimally with the argument ExperimentList,
potentially also with the arguments colData (see section below) and
sampleMap." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MultiAssayExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.12.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Quick Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/QuickStartMultiAssay.html">Quick Start Guide</a>
    </li>
    <li>
      <a href="../articles/MultiAssayExperiment.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/UsingHDF5Array.html">Using DelayedMatrix</a>
    </li>
    <li>
      <a href="../articles/extras/MAE-curatedMetagenomicData.html">MultiAssayExperiment in Practice</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Release Notes</li>
    <li>
      <a href="https://waldronlab.io/MultiAssayExperiment/news/news-1.12.html">Version 1.12.0</a>
    </li>
    <li>
      <a href="https://waldronlab.io/MultiAssayExperiment/news/news-1.10.html">Version 1.10.0</a>
    </li>
    <li>
      <a href="https://waldronlab.io/MultiAssayExperiment/news/news-1.8.html">Version 1.8.0</a>
    </li>
    <li>
      <a href="https://waldronlab.io/MultiAssayExperiment/news/news-1.7.html">Version 1.7.14</a>
    </li>
    <li>
      <a href="https://waldronlab.io/MultiAssayExperiment/news/news-1.1.html">Version 1.1.0</a>
    </li>
    <li>
      <a href="https://waldronlab.io/MultiAssayExperiment/news/news-0.101.html">Version 0.101.0</a>
    </li>
    <li>
      <a href="https://waldronlab.io/MultiAssayExperiment/news/news-0.99.html">Version 0.99.0</a>
    </li>
    <li>
      <a href="../news/index.html">Change log</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/waldronlab/MultiAssayExperiment">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>An integrative multi-assay class for experiment data</h1>
    
    <div class="hidden name"><code>MultiAssayExperiment-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>MultiAssayExperiment</code> class can be used to manage results of
diverse assays on a collection of specimen. Currently,  the class can handle
assays that are organized instances of
<code>SummarizedExperiment</code>,
<code>ExpressionSet</code>, <code>matrix</code>,
<code><a href='https://rdrr.io/pkg/RaggedExperiment/man/RaggedExperiment-class.html'>RaggedExperiment</a></code>
(inherits from <code>GRangesList</code>), and <code>RangedVcfStack</code>.
Create new <code>MultiAssayExperiment</code> instances with the homonymous
constructor, minimally with the argument <code><a href='ExperimentList.html'>ExperimentList</a></code>,
potentially also with the arguments <code>colData</code> (see section below) and
<code><a href='MultiAssayExperiment-methods.html'>sampleMap</a></code>.</p>
    </div>

    <pre class="usage"><span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>show</span>(<span class='no'>object</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>x</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='no'>x</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>updateObject</span>(<span class='no'>object</span>, <span class='no'>...</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>exportClass</span>(
  <span class='no'>object</span>,
  <span class='kw'>dir</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span>(),
  <span class='no'>fmt</span>,
  <span class='no'>ext</span>,
  <span class='kw'>match</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'><a href='https://rdrr.io/r/base/dimnames.html'>dimnames</a></span>(<span class='no'>x</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='no'>x</span>, <span class='no'>...</span>, <span class='kw'>sampleMap</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>mapFrom</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>assays</span>(<span class='no'>x</span>, <span class='no'>...</span>, <span class='kw'>withDimnames</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># S4 method for MultiAssayExperiment,missing</span>
<span class='fu'>assay</span>(<span class='no'>x</span>, <span class='no'>i</span>, <span class='no'>...</span>)

<span class='co'># S4 method for MultiAssayExperiment,numeric</span>
<span class='fu'>assay</span>(<span class='no'>x</span>, <span class='no'>i</span>, <span class='no'>...</span>)

<span class='co'># S4 method for MultiAssayExperiment,character</span>
<span class='fu'>assay</span>(<span class='no'>x</span>, <span class='no'>i</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>A <code>MultiAssayExperiment</code> object</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>A <code>MultiAssayExperiment</code> object</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments for supporting functions. See details.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical (default FALSE) whether to print extra messages</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p>character(1) A directory for saving exported data (default:
`tempdir()`)</p></td>
    </tr>
    <tr>
      <th>fmt</th>
      <td><p>character(1) or function() Either a format character atomic as
supported by `write.table` either ('csv', or 'tsv') or a function whose
first two arguments are 'object to save' and 'file location'</p></td>
    </tr>
    <tr>
      <th>ext</th>
      <td><p>character(1) A file extension supported by the format argument</p></td>
    </tr>
    <tr>
      <th>match</th>
      <td><p>logical(1) Whether to coerce the current object to a
'MatchedAssayExperiment' object (default: FALSE)</p></td>
    </tr>
    <tr>
      <th>sampleMap</th>
      <td><p><code>c</code> method: a <code>sampleMap</code> <code>list</code> or
<code>DataFrame</code> to guide merge</p></td>
    </tr>
    <tr>
      <th>mapFrom</th>
      <td><p>Either a <code>logical</code>, <code>character</code>, or <code>integer</code>
vector indicating the experiment(s) that have an identical colname order as
the experiment input(s)</p></td>
    </tr>
    <tr>
      <th>withDimnames</th>
      <td><p>logical (default TRUE) whether to return dimension names
included in the object</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>An integer or character scalar indicating the assay to return</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>MultiAssayExperiment</code> object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The dots (<code>...</code>) argument allows the user to specify additional
arguments in several instances.</p><ul>
<li><p>subsetting <strong>[</strong>: additional arguments sent to
    <a href='https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html'>findOverlaps</a>.</p></li>
<li><p>mergeReplicates: used to specify arguments for the <code>simplify</code>
    functional argument</p></li>
<li><p>assay: may contain withDimnames, which is forwarded to assays</p></li>
<li><p>combining <strong>c</strong>: compatible <code>MultiAssayExperiment</code> classes
    passed on to the <a href='ExperimentList-class.html'>ExperimentList</a> constructor, can be a
    <code>list</code>, List, or a series of named arguments. See
    the examples below.</p></li>
</ul>

    <h2 class="hasAnchor" id="methods-by-generic-"><a class="anchor" href="#methods-by-generic-"></a>Methods (by generic)</h2>

    
<ul>
<li><p><code>show</code>: Show method for a
<code>MultiAssayExperiment</code></p></li>
<li><p><code>length</code>: Get the length of ExperimentList</p></li>
<li><p><code>names</code>: Get the names of the ExperimentList</p></li>
<li><p><code>updateObject</code>: Update old serialized MultiAssayExperiment
objects to new API</p></li>
<li><p><code>exportClass</code>: Export data from class to a series of text
files</p></li>
<li><p><code>dimnames</code>: Get the dimension names
for a <code>MultiAssayExperiment</code> object</p></li>
<li><p><code>c</code>: Add a supported data class to the
<code>ExperimentList</code></p></li>
<li><p><code>assays</code>: Obtain a <code>SimpleList</code> of assay
data for all available experiments in the object</p></li>
<li><p><code>assay</code>: Convenience function for extracting the
assay of the first element (default) in the <code>ExperimentList</code>. A
<code>numeric</code> or <code>character</code> index can also be provided</p></li>
</ul>
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    

<dl class='dl-horizontal'>
<dt><code>ExperimentList</code></dt><dd><p>A <code><a href='ExperimentList.html'>ExperimentList</a></code> class object for
each assay dataset</p></dd>

<dt><code>colData</code></dt><dd><p>A <code>DataFrame</code> of all clinical/specimen data available
across experiments</p></dd>

<dt><code>sampleMap</code></dt><dd><p>A <code>DataFrame</code> of translatable identifiers
of samples and participants</p></dd>

<dt><code>metadata</code></dt><dd><p>Additional data describing the
<code>MultiAssayExperiment</code> object</p></dd>

<dt><code>drops</code></dt><dd><p>A metadata <code>list</code> of dropped information</p></dd>

</dl>
    <h2 class="hasAnchor" id="coldata"><a class="anchor" href="#coldata"></a>colData</h2>

    

<p>The <code>colData</code> slot is a collection of primary specimen data valid
across all experiments. This slot is strictly of class
<code>DataFrame</code> but arguments for the constructor function
allow arguments to be of class <code>data.frame</code> and subsequently coerced.</p>
    <h2 class="hasAnchor" id="experimentlist"><a class="anchor" href="#experimentlist"></a>ExperimentList</h2>

    

<p>The <code><a href='ExperimentList.html'>ExperimentList</a></code> slot is designed to contain results from
each experiment/assay. It contains a SimpleList.</p>
    <h2 class="hasAnchor" id="samplemap"><a class="anchor" href="#samplemap"></a>sampleMap</h2>

    

<p>The <code><a href='MultiAssayExperiment-methods.html'>sampleMap</a></code> contains a <code>DataFrame</code> of translatable
identifiers of samples and participants or biological units. Standard column
names of the sampleMap are "assay", "primary", and "colname".</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='MultiAssayExperiment-methods.html'>MultiAssayExperiment-methods</a> for slot modifying methods
    <a href='https://github.com/waldronlab/MultiAssayExperiment/wiki/MultiAssayExperiment-API'>MultiAssayExperiment API</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/example.html'>example</a></span>(<span class='st'>"MultiAssayExperiment"</span>)</div><div class='output co'>#&gt; 
#&gt; MltAsE&gt; ## Run the example ExperimentList
#&gt; MltAsE&gt; example("ExperimentList")
#&gt; 
#&gt; ExprmL&gt; ## Create an empty ExperimentList instance
#&gt; ExprmL&gt; ExperimentList()
#&gt; ExperimentList class object of length 0:
#&gt;  
#&gt; ExprmL&gt; ## Create array matrix and AnnotatedDataFrame to create an ExpressionSet class
#&gt; ExprmL&gt; arraydat &lt;- matrix(data = seq(101, length.out = 20), ncol = 4,
#&gt; ExprmL+     dimnames = list(
#&gt; ExprmL+         c("ENST00000294241", "ENST00000355076",
#&gt; ExprmL+         "ENST00000383706","ENST00000234812", "ENST00000383323"),
#&gt; ExprmL+         c("array1", "array2", "array3", "array4")
#&gt; ExprmL+     ))
#&gt; 
#&gt; ExprmL&gt; colDat &lt;- data.frame(slope53 = rnorm(4),
#&gt; ExprmL+     row.names = c("array1", "array2", "array3", "array4"))
#&gt; 
#&gt; ExprmL&gt; ## SummarizedExperiment constructor
#&gt; ExprmL&gt; exprdat &lt;- SummarizedExperiment::SummarizedExperiment(arraydat,
#&gt; ExprmL+     colData = colDat)
#&gt; 
#&gt; ExprmL&gt; ## Create a sample methylation dataset
#&gt; ExprmL&gt; methyldat &lt;- matrix(data = seq(1, length.out = 25), ncol = 5,
#&gt; ExprmL+     dimnames = list(
#&gt; ExprmL+         c("ENST00000355076", "ENST00000383706",
#&gt; ExprmL+           "ENST00000383323", "ENST00000234812", "ENST00000294241"),
#&gt; ExprmL+         c("methyl1", "methyl2", "methyl3",
#&gt; ExprmL+           "methyl4", "methyl5")
#&gt; ExprmL+     ))
#&gt; 
#&gt; ExprmL&gt; ## Create a sample RNASeqGene dataset
#&gt; ExprmL&gt; rnadat &lt;- matrix(
#&gt; ExprmL+     data = sample(c(46851, 5, 19, 13, 2197, 507,
#&gt; ExprmL+         84318, 126, 17, 21, 23979, 614), size = 20, replace = TRUE),
#&gt; ExprmL+     ncol = 4,
#&gt; ExprmL+     dimnames = list(
#&gt; ExprmL+         c("XIST", "RPS4Y1", "KDM5D", "ENST00000383323", "ENST00000234812"),
#&gt; ExprmL+         c("samparray1", "samparray2", "samparray3", "samparray4")
#&gt; ExprmL+     ))
#&gt; 
#&gt; ExprmL&gt; ## Create a mock RangedSummarizedExperiment from a data.frame
#&gt; ExprmL&gt; rangedat &lt;- data.frame(chr="chr2", start = 11:15, end = 12:16,
#&gt; ExprmL+     strand = c("+", "-", "+", "*", "."),
#&gt; ExprmL+     samp0 = c(0,0,1,1,1), samp1 = c(1,0,1,0,1), samp2 = c(0,1,0,1,0),
#&gt; ExprmL+     row.names = c(paste0("ENST", "00000", 135411:135414), "ENST00000383323"))
#&gt; 
#&gt; ExprmL&gt; rangeSE &lt;- SummarizedExperiment::makeSummarizedExperimentFromDataFrame(rangedat)
#&gt; 
#&gt; ExprmL&gt; ## Combine to a named list and call the ExperimentList constructor function
#&gt; ExprmL&gt; assayList &lt;- list(Affy = exprdat, Methyl450k = methyldat, RNASeqGene = rnadat,
#&gt; ExprmL+                 GISTIC = rangeSE)
#&gt; 
#&gt; ExprmL&gt; ## Use the ExperimentList constructor
#&gt; ExprmL&gt; ExpList &lt;- ExperimentList(assayList)
#&gt; 
#&gt; MltAsE&gt; ## Create sample maps for each experiment
#&gt; MltAsE&gt; exprmap &lt;- data.frame(
#&gt; MltAsE+     primary = c("Jack", "Jill", "Barbara", "Bob"),
#&gt; MltAsE+     colname = c("array1", "array2", "array3", "array4"),
#&gt; MltAsE+     stringsAsFactors = FALSE)
#&gt; 
#&gt; MltAsE&gt; methylmap &lt;- data.frame(
#&gt; MltAsE+     primary = c("Jack", "Jack", "Jill", "Barbara", "Bob"),
#&gt; MltAsE+     colname = c("methyl1", "methyl2", "methyl3", "methyl4", "methyl5"),
#&gt; MltAsE+     stringsAsFactors = FALSE)
#&gt; 
#&gt; MltAsE&gt; rnamap &lt;- data.frame(
#&gt; MltAsE+     primary = c("Jack", "Jill", "Bob", "Barbara"),
#&gt; MltAsE+     colname = c("samparray1", "samparray2", "samparray3", "samparray4"),
#&gt; MltAsE+     stringsAsFactors = FALSE)
#&gt; 
#&gt; MltAsE&gt; gistmap &lt;- data.frame(
#&gt; MltAsE+     primary = c("Jack", "Bob", "Jill"),
#&gt; MltAsE+     colname = c("samp0", "samp1", "samp2"),
#&gt; MltAsE+     stringsAsFactors = FALSE)
#&gt; 
#&gt; MltAsE&gt; ## Combine as a named list and convert to a DataFrame
#&gt; MltAsE&gt; maplist &lt;- list(Affy = exprmap, Methyl450k = methylmap,
#&gt; MltAsE+     RNASeqGene = rnamap, GISTIC = gistmap)
#&gt; 
#&gt; MltAsE&gt; ## Create a sampleMap
#&gt; MltAsE&gt; sampMap &lt;- listToMap(maplist)
#&gt; 
#&gt; MltAsE&gt; ## Create an example phenotype data
#&gt; MltAsE&gt; colDat &lt;- data.frame(sex = c("M", "F", "M", "F"), age = 38:41,
#&gt; MltAsE+     row.names = c("Jack", "Jill", "Bob", "Barbara"))
#&gt; 
#&gt; MltAsE&gt; ## Create a MultiAssayExperiment instance
#&gt; MltAsE&gt; mae &lt;- MultiAssayExperiment(experiments = ExpList, colData = colDat,
#&gt; MltAsE+     sampleMap = sampMap)</div><div class='input'>
<span class='co'>## Subsetting</span>
<span class='co'># Rows (i) Rows/Features in each experiment</span>
<span class='no'>mae</span>[<span class='fl'>1</span>, , ]</div><div class='output co'>#&gt; A MultiAssayExperiment object of 4 listed
#&gt;  experiments with user-defined names and respective classes.
#&gt;  Containing an ExperimentList class object of length 4:
#&gt;  [1] Affy: SummarizedExperiment with 1 rows and 4 columns
#&gt;  [2] Methyl450k: matrix with 1 rows and 5 columns
#&gt;  [3] RNASeqGene: matrix with 1 rows and 4 columns
#&gt;  [4] GISTIC: RangedSummarizedExperiment with 1 rows and 3 columns
#&gt; Features:
#&gt;  experiments() - obtain the ExperimentList instance
#&gt;  colData() - the primary/phenotype DFrame
#&gt;  sampleMap() - the sample availability DFrame
#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment
#&gt;  *Format() - convert into a long or wide DFrame
#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</div><div class='input'><span class='no'>mae</span>[<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>TRUE</span>, <span class='fl'>FALSE</span>), , ]</div><div class='output co'>#&gt; A MultiAssayExperiment object of 4 listed
#&gt;  experiments with user-defined names and respective classes.
#&gt;  Containing an ExperimentList class object of length 4:
#&gt;  [1] Affy: SummarizedExperiment with 3 rows and 4 columns
#&gt;  [2] Methyl450k: matrix with 3 rows and 5 columns
#&gt;  [3] RNASeqGene: matrix with 3 rows and 4 columns
#&gt;  [4] GISTIC: RangedSummarizedExperiment with 3 rows and 3 columns
#&gt; Features:
#&gt;  experiments() - obtain the ExperimentList instance
#&gt;  colData() - the primary/phenotype DFrame
#&gt;  sampleMap() - the sample availability DFrame
#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment
#&gt;  *Format() - convert into a long or wide DFrame
#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</div><div class='input'>
<span class='co'># Columns (j) Rows in colData</span>
<span class='no'>mae</span>[, <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='fu'>colData</span>(<span class='no'>mae</span>))[<span class='fl'>3</span>:<span class='fl'>2</span>],  ]</div><div class='output co'>#&gt; A MultiAssayExperiment object of 4 listed
#&gt;  experiments with user-defined names and respective classes.
#&gt;  Containing an ExperimentList class object of length 4:
#&gt;  [1] Affy: SummarizedExperiment with 5 rows and 2 columns
#&gt;  [2] Methyl450k: matrix with 5 rows and 2 columns
#&gt;  [3] RNASeqGene: matrix with 5 rows and 2 columns
#&gt;  [4] GISTIC: RangedSummarizedExperiment with 5 rows and 2 columns
#&gt; Features:
#&gt;  experiments() - obtain the ExperimentList instance
#&gt;  colData() - the primary/phenotype DFrame
#&gt;  sampleMap() - the sample availability DFrame
#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment
#&gt;  *Format() - convert into a long or wide DFrame
#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</div><div class='input'>
<span class='co'># Assays (k)</span>
<span class='no'>mae</span>[, , <span class='st'>"Affy"</span>]</div><div class='output co'>#&gt; <span class='message'>harmonizing input:</span>
#&gt; <span class='message'>  removing 12 sampleMap rows not in names(experiments)</span></div><div class='output co'>#&gt; A MultiAssayExperiment object of 1 listed
#&gt;  experiment with a user-defined name and respective class.
#&gt;  Containing an ExperimentList class object of length 1:
#&gt;  [1] Affy: SummarizedExperiment with 5 rows and 4 columns
#&gt; Features:
#&gt;  experiments() - obtain the ExperimentList instance
#&gt;  colData() - the primary/phenotype DFrame
#&gt;  sampleMap() - the sample availability DFrame
#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment
#&gt;  *Format() - convert into a long or wide DFrame
#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</div><div class='input'>
<span class='co'>## Complete cases (returns logical vector)</span>
<span class='no'>completes</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/complete.cases.html'>complete.cases</a></span>(<span class='no'>mae</span>)
<span class='no'>compMAE</span> <span class='kw'>&lt;-</span> <span class='no'>mae</span>[, <span class='no'>completes</span>, ]
<span class='no'>compMAE</span></div><div class='output co'>#&gt; A MultiAssayExperiment object of 4 listed
#&gt;  experiments with user-defined names and respective classes.
#&gt;  Containing an ExperimentList class object of length 4:
#&gt;  [1] Affy: SummarizedExperiment with 5 rows and 3 columns
#&gt;  [2] Methyl450k: matrix with 5 rows and 4 columns
#&gt;  [3] RNASeqGene: matrix with 5 rows and 3 columns
#&gt;  [4] GISTIC: RangedSummarizedExperiment with 5 rows and 3 columns
#&gt; Features:
#&gt;  experiments() - obtain the ExperimentList instance
#&gt;  colData() - the primary/phenotype DFrame
#&gt;  sampleMap() - the sample availability DFrame
#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment
#&gt;  *Format() - convert into a long or wide DFrame
#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</div><div class='input'><span class='fu'>colData</span>(<span class='no'>compMAE</span>)</div><div class='output co'>#&gt; DataFrame with 3 rows and 2 columns
#&gt;           sex       age
#&gt;      &lt;factor&gt; &lt;integer&gt;
#&gt; Jack        M        38
#&gt; Jill        F        39
#&gt; Bob         M        40</div><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/utils/example.html'>example</a></span>(<span class='st'>"MultiAssayExperiment"</span>)</div><div class='output co'>#&gt; 
#&gt; MltAsE&gt; ## Run the example ExperimentList
#&gt; MltAsE&gt; example("ExperimentList")
#&gt; 
#&gt; ExprmL&gt; ## Create an empty ExperimentList instance
#&gt; ExprmL&gt; ExperimentList()
#&gt; ExperimentList class object of length 0:
#&gt;  
#&gt; ExprmL&gt; ## Create array matrix and AnnotatedDataFrame to create an ExpressionSet class
#&gt; ExprmL&gt; arraydat &lt;- matrix(data = seq(101, length.out = 20), ncol = 4,
#&gt; ExprmL+     dimnames = list(
#&gt; ExprmL+         c("ENST00000294241", "ENST00000355076",
#&gt; ExprmL+         "ENST00000383706","ENST00000234812", "ENST00000383323"),
#&gt; ExprmL+         c("array1", "array2", "array3", "array4")
#&gt; ExprmL+     ))
#&gt; 
#&gt; ExprmL&gt; colDat &lt;- data.frame(slope53 = rnorm(4),
#&gt; ExprmL+     row.names = c("array1", "array2", "array3", "array4"))
#&gt; 
#&gt; ExprmL&gt; ## SummarizedExperiment constructor
#&gt; ExprmL&gt; exprdat &lt;- SummarizedExperiment::SummarizedExperiment(arraydat,
#&gt; ExprmL+     colData = colDat)
#&gt; 
#&gt; ExprmL&gt; ## Create a sample methylation dataset
#&gt; ExprmL&gt; methyldat &lt;- matrix(data = seq(1, length.out = 25), ncol = 5,
#&gt; ExprmL+     dimnames = list(
#&gt; ExprmL+         c("ENST00000355076", "ENST00000383706",
#&gt; ExprmL+           "ENST00000383323", "ENST00000234812", "ENST00000294241"),
#&gt; ExprmL+         c("methyl1", "methyl2", "methyl3",
#&gt; ExprmL+           "methyl4", "methyl5")
#&gt; ExprmL+     ))
#&gt; 
#&gt; ExprmL&gt; ## Create a sample RNASeqGene dataset
#&gt; ExprmL&gt; rnadat &lt;- matrix(
#&gt; ExprmL+     data = sample(c(46851, 5, 19, 13, 2197, 507,
#&gt; ExprmL+         84318, 126, 17, 21, 23979, 614), size = 20, replace = TRUE),
#&gt; ExprmL+     ncol = 4,
#&gt; ExprmL+     dimnames = list(
#&gt; ExprmL+         c("XIST", "RPS4Y1", "KDM5D", "ENST00000383323", "ENST00000234812"),
#&gt; ExprmL+         c("samparray1", "samparray2", "samparray3", "samparray4")
#&gt; ExprmL+     ))
#&gt; 
#&gt; ExprmL&gt; ## Create a mock RangedSummarizedExperiment from a data.frame
#&gt; ExprmL&gt; rangedat &lt;- data.frame(chr="chr2", start = 11:15, end = 12:16,
#&gt; ExprmL+     strand = c("+", "-", "+", "*", "."),
#&gt; ExprmL+     samp0 = c(0,0,1,1,1), samp1 = c(1,0,1,0,1), samp2 = c(0,1,0,1,0),
#&gt; ExprmL+     row.names = c(paste0("ENST", "00000", 135411:135414), "ENST00000383323"))
#&gt; 
#&gt; ExprmL&gt; rangeSE &lt;- SummarizedExperiment::makeSummarizedExperimentFromDataFrame(rangedat)
#&gt; 
#&gt; ExprmL&gt; ## Combine to a named list and call the ExperimentList constructor function
#&gt; ExprmL&gt; assayList &lt;- list(Affy = exprdat, Methyl450k = methyldat, RNASeqGene = rnadat,
#&gt; ExprmL+                 GISTIC = rangeSE)
#&gt; 
#&gt; ExprmL&gt; ## Use the ExperimentList constructor
#&gt; ExprmL&gt; ExpList &lt;- ExperimentList(assayList)
#&gt; 
#&gt; MltAsE&gt; ## Create sample maps for each experiment
#&gt; MltAsE&gt; exprmap &lt;- data.frame(
#&gt; MltAsE+     primary = c("Jack", "Jill", "Barbara", "Bob"),
#&gt; MltAsE+     colname = c("array1", "array2", "array3", "array4"),
#&gt; MltAsE+     stringsAsFactors = FALSE)
#&gt; 
#&gt; MltAsE&gt; methylmap &lt;- data.frame(
#&gt; MltAsE+     primary = c("Jack", "Jack", "Jill", "Barbara", "Bob"),
#&gt; MltAsE+     colname = c("methyl1", "methyl2", "methyl3", "methyl4", "methyl5"),
#&gt; MltAsE+     stringsAsFactors = FALSE)
#&gt; 
#&gt; MltAsE&gt; rnamap &lt;- data.frame(
#&gt; MltAsE+     primary = c("Jack", "Jill", "Bob", "Barbara"),
#&gt; MltAsE+     colname = c("samparray1", "samparray2", "samparray3", "samparray4"),
#&gt; MltAsE+     stringsAsFactors = FALSE)
#&gt; 
#&gt; MltAsE&gt; gistmap &lt;- data.frame(
#&gt; MltAsE+     primary = c("Jack", "Bob", "Jill"),
#&gt; MltAsE+     colname = c("samp0", "samp1", "samp2"),
#&gt; MltAsE+     stringsAsFactors = FALSE)
#&gt; 
#&gt; MltAsE&gt; ## Combine as a named list and convert to a DataFrame
#&gt; MltAsE&gt; maplist &lt;- list(Affy = exprmap, Methyl450k = methylmap,
#&gt; MltAsE+     RNASeqGene = rnamap, GISTIC = gistmap)
#&gt; 
#&gt; MltAsE&gt; ## Create a sampleMap
#&gt; MltAsE&gt; sampMap &lt;- listToMap(maplist)
#&gt; 
#&gt; MltAsE&gt; ## Create an example phenotype data
#&gt; MltAsE&gt; colDat &lt;- data.frame(sex = c("M", "F", "M", "F"), age = 38:41,
#&gt; MltAsE+     row.names = c("Jack", "Jill", "Bob", "Barbara"))
#&gt; 
#&gt; MltAsE&gt; ## Create a MultiAssayExperiment instance
#&gt; MltAsE&gt; mae &lt;- MultiAssayExperiment(experiments = ExpList, colData = colDat,
#&gt; MltAsE+     sampleMap = sampMap)</div><div class='input'>
<span class='co'>## Add an experiment</span>
<span class='no'>test1</span> <span class='kw'>&lt;-</span> <span class='no'>mae</span><span class='kw'>[[</span><span class='fl'>1L</span>]]
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>test1</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='fu'>colData</span>(<span class='no'>mae</span>))

<span class='co'>## Combine current MultiAssayExperiment with additional experiment</span>
<span class='co'>## (no sampleMap)</span>
<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='no'>mae</span>, <span class='kw'>newExperiment</span> <span class='kw'>=</span> <span class='no'>test1</span>)</div><div class='output co'>#&gt; A MultiAssayExperiment object of 5 listed
#&gt;  experiments with user-defined names and respective classes.
#&gt;  Containing an ExperimentList class object of length 5:
#&gt;  [1] Affy: SummarizedExperiment with 5 rows and 4 columns
#&gt;  [2] Methyl450k: matrix with 5 rows and 5 columns
#&gt;  [3] RNASeqGene: matrix with 5 rows and 4 columns
#&gt;  [4] GISTIC: RangedSummarizedExperiment with 5 rows and 3 columns
#&gt;  [5] newExperiment: SummarizedExperiment with 5 rows and 4 columns
#&gt; Features:
#&gt;  experiments() - obtain the ExperimentList instance
#&gt;  colData() - the primary/phenotype DFrame
#&gt;  sampleMap() - the sample availability DFrame
#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment
#&gt;  *Format() - convert into a long or wide DFrame
#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</div><div class='input'>
<span class='no'>test2</span> <span class='kw'>&lt;-</span> <span class='no'>mae</span><span class='kw'>[[</span><span class='fl'>3L</span>]]
<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='no'>mae</span>, <span class='kw'>newExp</span> <span class='kw'>=</span> <span class='no'>test2</span>, <span class='kw'>mapFrom</span> <span class='kw'>=</span> <span class='fl'>3L</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: Assuming column order in the data provided </span>
#&gt; <span class='warning'> matches the order in 'mapFrom' experiment(s) colnames</span></div><div class='output co'>#&gt; A MultiAssayExperiment object of 5 listed
#&gt;  experiments with user-defined names and respective classes.
#&gt;  Containing an ExperimentList class object of length 5:
#&gt;  [1] Affy: SummarizedExperiment with 5 rows and 4 columns
#&gt;  [2] Methyl450k: matrix with 5 rows and 5 columns
#&gt;  [3] RNASeqGene: matrix with 5 rows and 4 columns
#&gt;  [4] GISTIC: RangedSummarizedExperiment with 5 rows and 3 columns
#&gt;  [5] newExp: matrix with 5 rows and 4 columns
#&gt; Features:
#&gt;  experiments() - obtain the ExperimentList instance
#&gt;  colData() - the primary/phenotype DFrame
#&gt;  sampleMap() - the sample availability DFrame
#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment
#&gt;  *Format() - convert into a long or wide DFrame
#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#methods-by-generic-">Methods (by generic)</a></li>
      <li><a href="#slots">Slots</a></li>
      <li><a href="#coldata">colData</a></li>
      <li><a href="#experimentlist">ExperimentList</a></li>
      <li><a href="#samplemap">sampleMap</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Marcel Ramos, Levi Waldron.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


