<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MultiAssayExperiment: The Integrative Bioconductor Container • MultiAssayExperiment</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MultiAssayExperiment: The Integrative Bioconductor Container">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MultiAssayExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.9.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Quick Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/QuickStartMultiAssay.html">Quick Start Guide</a>
    </li>
    <li>
      <a href="../articles/MultiAssayExperiment.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/UsingHDF5Array.html">Using DelayedMatrix</a>
    </li>
    <li>
      <a href="../articles/extras/MAE-curatedMetagenomicData.html">MultiAssayExperiment in Practice</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MultiAssayExperiment: The Integrative Bioconductor Container</h1>
                        <h4 class="author">MultiAssay Special Interest Group</h4>
            
            <h4 class="date">February 12, 2019</h4>
      
      
      <div class="hidden name"><code>MultiAssayExperiment.Rmd</code></div>

    </div>

    
    
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"BiocManager"</span>))</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">    <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"MultiAssayExperiment"</span>)</a></code></pre></div>
<p>Loading the packages:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MultiAssayExperiment)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(GenomicRanges)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(SummarizedExperiment)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(RaggedExperiment)</a></code></pre></div>
</div>
<div id="a-brief-description" class="section level1">
<h1 class="hasAnchor">
<a href="#a-brief-description" class="anchor"></a>A Brief Description</h1>
<p><code>MultiAssayExperiment</code> offers a data structure for representing and analyzing multi-omics experiments: a biological analysis approach utilizing multiple types of observations, such as DNA mutations and abundance of RNA and proteins, in the same biological specimens.</p>
<div id="choosing-the-appropriate-data-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#choosing-the-appropriate-data-structure" class="anchor"></a>Choosing the appropriate data structure</h2>
<p>For assays with different numbers of rows and even columns, <code>MultiAssayExperiment</code> is recommended. For sets of assays with the same information across all rows (e.g., genes or genomic ranges), <code>SummarizedExperiment</code> is the recommended data structure.</p>
</div>
</div>
<div id="overview-of-the-multiassayexperiment-class" class="section level1">
<h1 class="hasAnchor">
<a href="#overview-of-the-multiassayexperiment-class" class="anchor"></a>Overview of the <code>MultiAssayExperiment</code> class</h1>
<p>Here is an overview of the class and its constructors and extractors:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">empty &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment.html">MultiAssayExperiment</a></span>()</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">empty</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 0 listed
##  experiments with no user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 0:  
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">slotNames</span>(empty)</a></code></pre></div>
<pre><code>## [1] "ExperimentList" "colData"        "sampleMap"      "drops"         
## [5] "metadata"</code></pre>
<div id="components-of-the-multiassayexperiment" class="section level2">
<h2 class="hasAnchor">
<a href="#components-of-the-multiassayexperiment" class="anchor"></a>Components of the <code>MultiAssayExperiment</code>
</h2>
<div id="experimentlist-experimental-data" class="section level3">
<h3 class="hasAnchor">
<a href="#experimentlist-experimental-data" class="anchor"></a><code>ExperimentList</code>: experimental data</h3>
<p>The <code>ExperimentList</code> slot and class is the container workhorse for the <code>MultiAssayExperiment</code> class. It contains all the experimental data. It inherits from class <code><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/SimpleList-class">S4Vectors::SimpleList</a></code> with one element/component per data type.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(<span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">experiments</a></span>(empty)) <span class="co"># ExperimentList</span></a></code></pre></div>
<pre><code>## [1] "ExperimentList"
## attr(,"package")
## [1] "MultiAssayExperiment"</code></pre>
<p>The elements of the <code>ExperimentList</code> can contain <strong>ID-based</strong> and <strong>range-based</strong> data. Requirements for all classes in the <code>ExperimentList</code> are listed in the API.</p>
<p>The following base and Bioconductor classes are known to work as elements of the ExperimentList:</p>
<ul>
<li><p><code><a href="https://www.rdocumentation.org/packages/base/topics/matrix">base::matrix</a></code>: the base class, can be used for ID-based datasets such as gene expression summarized per-gene, microRNA, metabolomics, or microbiome data.</p></li>
<li><p><code><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class">SummarizedExperiment::SummarizedExperiment</a></code>: A richer representation compared to a ordinary matrix of ID-based datasets capable of storing additional assay- level metadata.</p></li>
<li><p><code><a href="https://www.rdocumentation.org/packages/Biobase/topics/class.ExpressionSet">Biobase::ExpressionSet</a></code>: A <em>legacy</em> representation of ID-based datasets, supported for convenience and supplanted by <code>SummarizedExperiment</code>.</p></li>
<li><p><code><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class">SummarizedExperiment::RangedSummarizedExperiment</a></code>: For rectangular range-based datasets, one set of genomic ranges are assayed for multiple samples. It can be used for gene expression, methylation, or other data types that refer to genomic positions.</p></li>
<li><p><code><a href="https://www.rdocumentation.org/packages/RaggedExperiment/topics/RaggedExperiment-class">RaggedExperiment::RaggedExperiment</a></code>: For range-based datasets, such as copy number and mutation data, the <code>RaggedExperiment</code> class can be used to represent measurements by genomic positions.</p></li>
</ul>
<div id="class-requirements-within-experimentlist-container" class="section level4">
<h4 class="hasAnchor">
<a href="#class-requirements-within-experimentlist-container" class="anchor"></a>Class requirements within <code>ExperimentList</code> container</h4>
<p>See the <a href="#application-programming-interface-api">API section</a> for details on requirements for using other data classes. In general, data classes meeting minimum requirements, including support for square bracket <code>[</code> subsetting and <code><a href="https://www.rdocumentation.org/packages/base/topics/dimnames">dimnames()</a></code> will work by default.</p>
<p>The datasets contained in elements of the <code>ExperimentList</code> can have:</p>
<ul>
<li>column names (required)</li>
<li>row names (optional)</li>
</ul>
<p>The column names correspond to samples, and are used to match assay data to specimen metadata stored in <code>colData</code>.</p>
<p>The row names can correspond to a variety of features in the data including but not limited to gene names, probe IDs, proteins, and named ranges. Note that the existence of “row” names does <em>not</em> mean the data must be rectangular or matrix-like.</p>
<p>Classes contained in the <code>ExperimentList</code> must support the following list of methods:</p>
<ul>
<li>
<code>[</code>: single square bracket subsetting, with a single comma. It is assumed that values before the comma subset rows, and values after the comma subset columns.</li>
<li>
<code><a href="https://www.rdocumentation.org/packages/base/topics/dimnames">dimnames()</a></code> : corresponding to features (such as genes, proteins, etc.) and experimental samples</li>
<li>
<code><a href="https://www.rdocumentation.org/packages/base/topics/dim">dim()</a></code>: returns a vector of the number of rows and number of columns</li>
</ul>
</div>
</div>
<div id="coldata-primary-data" class="section level3">
<h3 class="hasAnchor">
<a href="#coldata-primary-data" class="anchor"></a><code>colData</code>: primary data</h3>
<p>The <code>MultiAssayExperiment</code> keeps one set of “primary” metadata that describes the ‘biological unit’ which can refer to specimens, experimental subjects, patients, etc. In this vignette, we will refer to each experimental subject as a <em>patient</em>.</p>
<div id="coldata-slot-requirements" class="section level4">
<h4 class="hasAnchor">
<a href="#coldata-slot-requirements" class="anchor"></a><code>colData</code> slot requirements</h4>
<p>The <code>colData</code> dataset should be of class <code>DataFrame</code> but can accept a <code>data.frame</code> class object that will be coerced.</p>
<p>In order to relate metadata of the biological unit, the row names of the <code>colData</code> dataset must contain patient identifiers.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">patient.data &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">sex=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"M"</span>, <span class="st">"F"</span>, <span class="st">"M"</span>, <span class="st">"F"</span>),</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">    <span class="dt">age=</span><span class="dv">38</span><span class="op">:</span><span class="dv">41</span>,</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">    <span class="dt">row.names=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Jack"</span>, <span class="st">"Jill"</span>, <span class="st">"Bob"</span>, <span class="st">"Barbara"</span>))</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">patient.data</a></code></pre></div>
<pre><code>##         sex age
## Jack      M  38
## Jill      F  39
## Bob       M  40
## Barbara   F  41</code></pre>
<p>Key points:</p>
<ul>
<li>one row of <code>colData</code> <em>can</em> map to zero, one, or more columns in any <code>ExperimentList</code> element</li>
<li>each row of <code>colData</code> <em>must</em> map to at least one column in at least one <code>ExperimentList</code> element.</li>
<li>each column of each <code>ExperimentList</code> element <em>must</em> map to <em>exactly</em> one row of <code>colData</code>.</li>
</ul>
<p>These relationships are defined by the <a href="#sampleMap">sampleMap</a>.</p>
</div>
<div id="note-on-the-flexibility-of-the-dataframe" class="section level4">
<h4 class="hasAnchor">
<a href="#note-on-the-flexibility-of-the-dataframe" class="anchor"></a>Note on the flexibility of the <code>DataFrame</code>
</h4>
<p>For many typical purposes the <code>DataFrame</code> and <code>data.frame</code> behave equivalently; but the <code>Dataframe</code> is more flexible as it allows any vector-like data type to be stored in its columns. The flexibility of the <code>DataFrame</code> permits, for example, storing multiple dose-response values for a single cell line, even if the number of doses and responses is not consistent across all cell lines. Doses could be stored in one column of <code>colData</code> as a <code>SimpleList</code>, and responses in another column, also as a <code>SimpleList</code>. Or, dose-response values could be stored in a single column of <code>colData</code> as a two-column matrix for each cell line.</p>
</div>
</div>
<div id="sampleMap" class="section level3">
<h3 class="hasAnchor">
<a href="#sampleMap" class="anchor"></a><code>sampleMap</code>: relating <code>colData</code> to multiple assays</h3>
<p>The <code>sampleMap</code> is a <code>DataFrame</code> that relates the “primary” data (<code>colData</code>) to the experimental assays:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(<span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">sampleMap</a></span>(empty)) <span class="co"># DataFrame</span></a></code></pre></div>
<pre><code>## [1] "DataFrame"
## attr(,"package")
## [1] "S4Vectors"</code></pre>
<p>The <code>sampleMap</code> provides an unambiguous map from every experimental observation to <em>one and only one</em> row in <code>colData</code>. It is, however, permissible for a row of <code>colData</code> to be associated with multiple experimental observations or no observations at all. In other words, there is a “many-to-one” mapping from experimental observations to rows of <code>colData</code>, and a “one-to-any-number” mapping from rows of <code>colData</code> to experimental observations.</p>
<div id="samplemap-structure" class="section level4">
<h4 class="hasAnchor">
<a href="#samplemap-structure" class="anchor"></a><code>sampleMap</code> structure</h4>
<p>The <code>sampleMap</code> has three columns, with the following column names:</p>
<ol style="list-style-type: decimal">
<li><p><strong>assay</strong> provides the names of the different experiments / assays performed. These are user-defined, with the only requirement that the names of the <code>ExperimentList</code>, where the experimental assays are stored, must be contained in this column.</p></li>
<li><p><strong>primary</strong> provides the “primary” sample names. All values in this column must also be present in the rownames of <code><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData(MultiAssayExperiment)</a></code>. In this example, allowable values in this column are “Jack”, “Jill”, “Barbara”, and “Bob”.</p></li>
<li><p><strong>colname</strong> provides the sample names used by experimental datasets, which in practice are often different than the primary sample names. For each assay, all column names must be found in this column. Otherwise, those assays would be orphaned: it would be impossible to match them up to samples in the overall experiment. As mentioned above, duplicate values are allowed, to represent replicates with the same overall experiment-level annotation.</p></li>
</ol>
<p>This design is motivated by the following situations:</p>
<ol style="list-style-type: decimal">
<li>It allows flexibility for any amount of technical replication and biological replication (such as tumor and matched normal for a single patient) of individual assays.</li>
<li>It allows missing observations (such as RNA-seq performed only for some of the patients).</li>
<li>It allows the use of different identifiers to be used for patients / specimens and for each assay. These different identifiers are matched unambiguously, and consistency between them is maintained during subsetting and re-ordering.</li>
</ol>
<div id="instances-where-samplemap-isnt-provided" class="section level5">
<h5 class="hasAnchor">
<a href="#instances-where-samplemap-isnt-provided" class="anchor"></a>Instances where <code>sampleMap</code> isn’t provided</h5>
<p>If each assay uses the same colnames (i.e., if the same sample identifiers are used for each experiment), a simple list of these datasets is sufficient for the <code>MultiAssayExperiment</code> constructor function. It is not necessary for them to have the same rownames or colnames:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">exprss1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(<span class="dv">16</span>), <span class="dt">ncol =</span> <span class="dv">4</span>,</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">        <span class="dt">dimnames =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"ENST00000%i"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="dv">288754</span><span class="op">:</span><span class="dv">290000</span>, <span class="dv">4</span>)),</a>
<a class="sourceLine" id="cb13-3" data-line-number="3">                <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Jack"</span>, <span class="st">"Jill"</span>, <span class="st">"Bob"</span>, <span class="st">"Bobby"</span>)))</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">exprss2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(<span class="dv">12</span>), <span class="dt">ncol =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb13-5" data-line-number="5">        <span class="dt">dimnames =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"ENST00000%i"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="dv">288754</span><span class="op">:</span><span class="dv">290000</span>, <span class="dv">4</span>)),</a>
<a class="sourceLine" id="cb13-6" data-line-number="6">                <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Jack"</span>, <span class="st">"Jane"</span>, <span class="st">"Bob"</span>)))</a>
<a class="sourceLine" id="cb13-7" data-line-number="7">doubleExp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"methyl 2k"</span>  =<span class="st"> </span>exprss1, <span class="st">"methyl 3k"</span> =<span class="st"> </span>exprss2)</a>
<a class="sourceLine" id="cb13-8" data-line-number="8">simpleMultiAssay &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment.html">MultiAssayExperiment</a></span>(<span class="dt">experiments=</span>doubleExp)</a>
<a class="sourceLine" id="cb13-9" data-line-number="9">simpleMultiAssay</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 2: 
##  [1] methyl 2k: matrix with 4 rows and 4 columns 
##  [2] methyl 3k: matrix with 4 rows and 3 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>In the above example, the user did not provide the <code>colData</code> argument so the constructor function filled it with an empty <code>DataFrame</code>:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(simpleMultiAssay)</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 0 columns</code></pre>
<p>But the <code>colData</code> can be provided. Here, note that any assay sample (column) that cannot be mapped to a corresponding row in the provided <code>colData</code> gets dropped. This is part of ensuring internal validity of the <code>MultiAssayExperiment</code>.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">simpleMultiAssay2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment.html">MultiAssayExperiment</a></span>(<span class="dt">experiments=</span>doubleExp,</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">                                          <span class="dt">colData=</span>patient.data)</a></code></pre></div>
<pre><code>## Warning in .sampleMapFromData(colData, experiments): Data from rows:
##  NA - Bobby
##  NA - Jane
## dropped due to missing phenotype data</code></pre>
<pre><code>## harmonizing input:
##   removing 1 colData rownames not in sampleMap 'primary'</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">simpleMultiAssay2</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 2: 
##  [1] methyl 2k: matrix with 4 rows and 3 columns 
##  [2] methyl 3k: matrix with 4 rows and 2 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(simpleMultiAssay2)</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 2 columns
##           sex       age
##      &lt;factor&gt; &lt;integer&gt;
## Jack        M        38
## Jill        F        39
## Bob         M        40</code></pre>
</div>
</div>
</div>
<div id="metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata" class="anchor"></a>metadata</h3>
<p>Metadata can be added at different levels of the <code>MultiAssayExperiment</code>.</p>
<p>Can be of <em>ANY</em> class, for storing study-wide metadata, such as citation information. For an empty <code>MultiAssayExperiment</code> object, it is NULL.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(<span class="kw">metadata</span>(empty)) <span class="co"># NULL (class "ANY")</span></a></code></pre></div>
<pre><code>## [1] "list"</code></pre>
<p>At the <code>ExperimentList</code> level, the <code>metadata</code> function would allow the user to enter metadata as a <code>list</code>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw">metadata</span>(<span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">experiments</a></span>(empty))</a></code></pre></div>
<pre><code>## list()</code></pre>
<p>At the individual assay level, certain classes <em>may</em> support metadata, for example, <code>metadata</code> and <code>mcols</code> for a <code>SummarizedExperiment</code>. It is recommended to use <code>metadata</code> at the <code>ExperimentList</code> level.</p>
<p style="text-align: right;">
<a href="#header">back to top</a>
</p>
</div>
</div>
</div>
<div id="creating-a-multiassayexperiment-object-a-rich-example" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-a-multiassayexperiment-object-a-rich-example" class="anchor"></a>Creating a <code>MultiAssayExperiment</code> object: a rich example</h1>
<p>In this section we demonstrate all core supported data classes, using different sample ID conventions for each assay, with primary <code>colData</code>. The some supported classes such as, <code>matrix</code>, <code>SummarizedExperiment</code>, and <code>RangedSummarizedExperiment</code>.</p>
<div id="create-toy-datasets-demonstrating-all-supported-data-types" class="section level2">
<h2 class="hasAnchor">
<a href="#create-toy-datasets-demonstrating-all-supported-data-types" class="anchor"></a>Create toy datasets demonstrating all supported data types</h2>
<p>We have three matrix-like datasets. First, let’s represent expression data as a <code>SummarizedExperiment</code>:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">(arraydat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">101</span>, <span class="dv">108</span>), <span class="dt">ncol=</span><span class="dv">4</span>,</a>
<a class="sourceLine" id="cb28-2" data-line-number="2">    <span class="dt">dimnames=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ENST00000294241"</span>, <span class="st">"ENST00000355076"</span>),</a>
<a class="sourceLine" id="cb28-3" data-line-number="3">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"array1"</span>, <span class="st">"array2"</span>, <span class="st">"array3"</span>, <span class="st">"array4"</span>))))</a></code></pre></div>
<pre><code>##                 array1 array2 array3 array4
## ENST00000294241    101    103    105    107
## ENST00000355076    102    104    106    108</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1">coldat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">slope53=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(<span class="dv">4</span>),</a>
<a class="sourceLine" id="cb30-2" data-line-number="2">    <span class="dt">row.names=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"array1"</span>, <span class="st">"array2"</span>, <span class="st">"array3"</span>, <span class="st">"array4"</span>))</a>
<a class="sourceLine" id="cb30-3" data-line-number="3"></a>
<a class="sourceLine" id="cb30-4" data-line-number="4">exprdat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class">SummarizedExperiment</a></span>(arraydat, <span class="dt">colData=</span>coldat)</a>
<a class="sourceLine" id="cb30-5" data-line-number="5">exprdat</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 2 4 
## metadata(0):
## assays(1): ''
## rownames(2): ENST00000294241 ENST00000355076
## rowData names(0):
## colnames(4): array1 array2 array3 array4
## colData names(1): slope53</code></pre>
<p>The following map matches <code>colData</code> sample names to <code>exprdata</code> sample names. Note that row orders aren’t initially matched up, and this is OK.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">(exprmap &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">primary=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(patient.data)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">3</span>)],</a>
<a class="sourceLine" id="cb32-2" data-line-number="2">                       <span class="dt">colname=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"array1"</span>, <span class="st">"array2"</span>, <span class="st">"array3"</span>, <span class="st">"array4"</span>),</a>
<a class="sourceLine" id="cb32-3" data-line-number="3">                       <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>))</a></code></pre></div>
<pre><code>##   primary colname
## 1    Jack  array1
## 2    Jill  array2
## 3 Barbara  array3
## 4     Bob  array4</code></pre>
<p>Now methylation data, which we will represent as a <code>matrix</code>. It uses gene identifiers also, but measures a partially overlapping set of genes. Now, let’s store this as a simple matrix which can contains a replicate for one of the patients.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1">(methyldat &lt;-</a>
<a class="sourceLine" id="cb34-2" data-line-number="2"><span class="st">   </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">ncol=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb34-3" data-line-number="3">          <span class="dt">dimnames=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ENST00000355076"</span>, <span class="st">"ENST00000383706"</span>),</a>
<a class="sourceLine" id="cb34-4" data-line-number="4">                        <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"methyl1"</span>, <span class="st">"methyl2"</span>, <span class="st">"methyl3"</span>,</a>
<a class="sourceLine" id="cb34-5" data-line-number="5">                          <span class="st">"methyl4"</span>, <span class="st">"methyl5"</span>))))</a></code></pre></div>
<pre><code>##                 methyl1 methyl2 methyl3 methyl4 methyl5
## ENST00000355076       1       3       5       7       9
## ENST00000383706       2       4       6       8      10</code></pre>
<p>The following map matches <code>colData</code> sample names to <code>methyldat</code> sample names.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1">(methylmap &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">primary =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Jack"</span>, <span class="st">"Jack"</span>, <span class="st">"Jill"</span>, <span class="st">"Barbara"</span>, <span class="st">"Bob"</span>),</a>
<a class="sourceLine" id="cb36-2" data-line-number="2">    <span class="dt">colname =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"methyl1"</span>, <span class="st">"methyl2"</span>, <span class="st">"methyl3"</span>, <span class="st">"methyl4"</span>, <span class="st">"methyl5"</span>),</a>
<a class="sourceLine" id="cb36-3" data-line-number="3">    <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>))</a></code></pre></div>
<pre><code>##   primary colname
## 1    Jack methyl1
## 2    Jack methyl2
## 3    Jill methyl3
## 4 Barbara methyl4
## 5     Bob methyl5</code></pre>
<p>Now we have a microRNA platform, which has no common identifiers with the other datasets, and which we also represent as a <code>matrix</code>. It is also missing data for “Jill”. We will use the same sample naming convention as we did for arrays.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1">(microdat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dv">201</span><span class="op">:</span><span class="dv">212</span>, <span class="dt">ncol=</span><span class="dv">3</span>,</a>
<a class="sourceLine" id="cb38-2" data-line-number="2">                    <span class="dt">dimnames=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"hsa-miR-21"</span>, <span class="st">"hsa-miR-191"</span>,</a>
<a class="sourceLine" id="cb38-3" data-line-number="3">                                    <span class="st">"hsa-miR-148a"</span>, <span class="st">"hsa-miR148b"</span>),</a>
<a class="sourceLine" id="cb38-4" data-line-number="4">                                  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"micro1"</span>, <span class="st">"micro2"</span>, <span class="st">"micro3"</span>))))</a></code></pre></div>
<pre><code>##              micro1 micro2 micro3
## hsa-miR-21      201    205    209
## hsa-miR-191     202    206    210
## hsa-miR-148a    203    207    211
## hsa-miR148b     204    208    212</code></pre>
<p>And the following map matches <code>colData</code> sample names to <code>microdat</code> sample names.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">(micromap &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">primary =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Jack"</span>, <span class="st">"Barbara"</span>, <span class="st">"Bob"</span>),</a>
<a class="sourceLine" id="cb40-2" data-line-number="2">    <span class="dt">colname =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"micro1"</span>, <span class="st">"micro2"</span>, <span class="st">"micro3"</span>), <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>))</a></code></pre></div>
<pre><code>##   primary colname
## 1    Jack  micro1
## 2 Barbara  micro2
## 3     Bob  micro3</code></pre>
<p>Finally, we create a dataset of class <code>RangedSummarizedExperiment</code>:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1">nrows &lt;-<span class="st"> </span><span class="dv">5</span>; ncols &lt;-<span class="st"> </span><span class="dv">4</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2">counts &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(nrows <span class="op">*</span><span class="st"> </span>ncols, <span class="dv">1</span>, <span class="fl">1e4</span>), nrows)</a>
<a class="sourceLine" id="cb42-3" data-line-number="3">rowRanges &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicRanges/topics/GRanges-class">GRanges</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">2</span>, nrows <span class="op">-</span><span class="st"> </span><span class="dv">2</span>)),</a>
<a class="sourceLine" id="cb42-4" data-line-number="4">    <span class="kw">IRanges</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">floor</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(nrows, <span class="fl">1e5</span>, <span class="fl">1e6</span>)), <span class="dt">width=</span><span class="dv">100</span>),</a>
<a class="sourceLine" id="cb42-5" data-line-number="5">    <span class="dt">strand=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"+"</span>, <span class="st">"-"</span>), nrows, <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb42-6" data-line-number="6">    <span class="dt">feature_id=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"ID</span><span class="ch">\\</span><span class="st">%03d"</span>, <span class="dv">1</span><span class="op">:</span>nrows))</a>
<a class="sourceLine" id="cb42-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(rowRanges) &lt;-<span class="st"> </span>letters[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb42-8" data-line-number="8">colData &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/DataFrame-class">DataFrame</a></span>(<span class="dt">Treatment=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ChIP"</span>, <span class="st">"Input"</span>), <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb42-9" data-line-number="9">    <span class="dt">row.names=</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"mysnparray1"</span>, <span class="st">"mysnparray2"</span>, <span class="st">"mysnparray3"</span>, <span class="st">"mysnparray4"</span>))</a>
<a class="sourceLine" id="cb42-10" data-line-number="10">rse &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class">SummarizedExperiment</a></span>(<span class="dt">assays=</span><span class="kw">SimpleList</span>(<span class="dt">counts=</span>counts),</a>
<a class="sourceLine" id="cb42-11" data-line-number="11">    <span class="dt">rowRanges=</span>rowRanges, <span class="dt">colData=</span>colData)</a></code></pre></div>
<p>And we map the <code>colData</code> samples to the <code>RangedSummarizedExperiment</code>:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1">(rangemap &lt;-</a>
<a class="sourceLine" id="cb43-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">primary =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Jack"</span>, <span class="st">"Jill"</span>, <span class="st">"Bob"</span>, <span class="st">"Barbara"</span>),</a>
<a class="sourceLine" id="cb43-3" data-line-number="3">    <span class="dt">colname =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"mysnparray1"</span>, <span class="st">"mysnparray2"</span>, <span class="st">"mysnparray3"</span>, <span class="st">"mysnparray4"</span>),</a>
<a class="sourceLine" id="cb43-4" data-line-number="4">        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>))</a></code></pre></div>
<pre><code>##   primary     colname
## 1    Jack mysnparray1
## 2    Jill mysnparray2
## 3     Bob mysnparray3
## 4 Barbara mysnparray4</code></pre>
</div>
<div id="samplemap-creation" class="section level2">
<h2 class="hasAnchor">
<a href="#samplemap-creation" class="anchor"></a><code>sampleMap</code> creation</h2>
<p>The <code>MultiAssayExperiment</code> constructor function can create the <code>sampleMap</code> automatically if a single naming convention is used, but in this example it cannot because we used platform-specific sample identifiers (e.g. mysnparray1, etc). So we must provide an ID map that matches the samples of each experiment back to the <code>colData</code>, as a three-column <code>data.frame</code> or <code>DataFrame</code> with three columns named “assay”, primary“, and”colname". Here we start with a list:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1">listmap &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(exprmap, methylmap, micromap, rangemap)</a>
<a class="sourceLine" id="cb45-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(listmap) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Affy"</span>, <span class="st">"Methyl 450k"</span>, <span class="st">"Mirna"</span>, <span class="st">"CNV gistic"</span>)</a>
<a class="sourceLine" id="cb45-3" data-line-number="3">listmap</a></code></pre></div>
<pre><code>## $Affy
##   primary colname
## 1    Jack  array1
## 2    Jill  array2
## 3 Barbara  array3
## 4     Bob  array4
## 
## $`Methyl 450k`
##   primary colname
## 1    Jack methyl1
## 2    Jack methyl2
## 3    Jill methyl3
## 4 Barbara methyl4
## 5     Bob methyl5
## 
## $Mirna
##   primary colname
## 1    Jack  micro1
## 2 Barbara  micro2
## 3     Bob  micro3
## 
## $`CNV gistic`
##   primary     colname
## 1    Jack mysnparray1
## 2    Jill mysnparray2
## 3     Bob mysnparray3
## 4 Barbara mysnparray4</code></pre>
<p>and use the convenience function <code>listToMap</code> to convert the list of <code>data.frame</code> objects to a valid object for the <code>sampleMap</code>:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1">dfmap &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mapToList.html">listToMap</a></span>(listmap)</a>
<a class="sourceLine" id="cb47-2" data-line-number="2">dfmap</a></code></pre></div>
<pre><code>## DataFrame with 16 rows and 3 columns
##           assay     primary     colname
##        &lt;factor&gt; &lt;character&gt; &lt;character&gt;
## 1          Affy        Jack      array1
## 2          Affy        Jill      array2
## 3          Affy     Barbara      array3
## 4          Affy         Bob      array4
## 5   Methyl 450k        Jack     methyl1
## ...         ...         ...         ...
## 12        Mirna         Bob      micro3
## 13   CNV gistic        Jack mysnparray1
## 14   CNV gistic        Jill mysnparray2
## 15   CNV gistic         Bob mysnparray3
## 16   CNV gistic     Barbara mysnparray4</code></pre>
<p>Note, <code>dfmap</code> can be reverted to a list with another provided function:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1"><span class="kw"><a href="../reference/mapToList.html">mapToList</a></span>(dfmap, <span class="st">"assay"</span>)</a></code></pre></div>
</div>
<div id="experimental-data-as-a-list" class="section level2">
<h2 class="hasAnchor">
<a href="#experimental-data-as-a-list" class="anchor"></a>Experimental data as a <code><a href="https://www.rdocumentation.org/packages/base/topics/list">list()</a></code>
</h2>
<p>Create an named list of experiments for the <code>MultiAssayExperiment</code> function. All of these names must be found within in the third column of <code>dfmap</code>:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" data-line-number="1">objlist &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"Affy"</span> =<span class="st"> </span>exprdat, <span class="st">"Methyl 450k"</span> =<span class="st"> </span>methyldat,</a>
<a class="sourceLine" id="cb50-2" data-line-number="2">    <span class="st">"Mirna"</span> =<span class="st"> </span>microdat, <span class="st">"CNV gistic"</span> =<span class="st"> </span>rse)</a></code></pre></div>
</div>
<div id="creation-of-the-multiassayexperiment-class-object" class="section level2">
<h2 class="hasAnchor">
<a href="#creation-of-the-multiassayexperiment-class-object" class="anchor"></a>Creation of the <code>MultiAssayExperiment</code> class object</h2>
<p>We recommend using the <code>MultiAssayExperiment</code> constructor function:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1">myMultiAssay &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment.html">MultiAssayExperiment</a></span>(objlist, patient.data, dfmap)</a>
<a class="sourceLine" id="cb51-2" data-line-number="2">myMultiAssay</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 5 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>The following extractor functions can be used to get extract data from the object:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">experiments</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 5 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns</code></pre>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##              sex       age
##         &lt;factor&gt; &lt;integer&gt;
## Jack           M        38
## Jill           F        39
## Bob            M        40
## Barbara        F        41</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">sampleMap</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## DataFrame with 16 rows and 3 columns
##           assay     primary     colname
##        &lt;factor&gt; &lt;character&gt; &lt;character&gt;
## 1          Affy        Jack      array1
## 2          Affy        Jill      array2
## 3          Affy     Barbara      array3
## 4          Affy         Bob      array4
## 5   Methyl 450k        Jack     methyl1
## ...         ...         ...         ...
## 12        Mirna         Bob      micro3
## 13   CNV gistic        Jack mysnparray1
## 14   CNV gistic        Jill mysnparray2
## 15   CNV gistic         Bob mysnparray3
## 16   CNV gistic     Barbara mysnparray4</code></pre>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1"><span class="kw">metadata</span>(myMultiAssay)</a></code></pre></div>
<pre><code>## list()</code></pre>
<p>Note that the <code>ExperimentList</code> class extends the <code>SimpleList</code> class to add some validity checks specific to <code>MultiAssayExperiment</code>. It can be used like a list.</p>
</div>
<div id="helper-function-to-create-a-multiassayexperiment-object" class="section level2">
<h2 class="hasAnchor">
<a href="#helper-function-to-create-a-multiassayexperiment-object" class="anchor"></a>Helper function to create a <code>MultiAssayExperiment</code> object</h2>
<p>The <code>prepMultiAssay</code> function helps diagnose common problems when creating a <code>MultiAssayExperiment</code> object. It provides error messages and/or warnings in instances where names (either <code>colnames</code> or <code>ExperimentList</code> element names) are inconsistent with those found in the sampleMap. Input arguments are the same as those in the <code>MultiAssayExperiment</code> (i.e., <code>ExperimentList</code>, <code>colData</code>, <code>sampleMap</code>). The resulting output of the <code>prepMultiAssay</code> function is a list of inputs including a “metadata$drops” element for names that were not able to be matched.</p>
<p>Instances where <code>ExperimentList</code> is created without names will prompt an error from <code>prepMultiAssay</code>. Named <code>ExperimentList</code> elements are essential for checks in <code>MultiAssayExperiment</code>.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" data-line-number="1">objlist3 &lt;-<span class="st"> </span>objlist</a>
<a class="sourceLine" id="cb61-2" data-line-number="2">(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(objlist3) &lt;-<span class="st"> </span><span class="ot">NULL</span>)</a></code></pre></div>
<pre><code>## NULL</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/try">try</a></span>(<span class="kw"><a href="../reference/prepMultiAssay.html">prepMultiAssay</a></span>(objlist3, patient.data, dfmap)<span class="op">$</span>experiments,</a>
<a class="sourceLine" id="cb63-2" data-line-number="2">    <span class="dt">outFile =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/showConnections">stdout</a></span>())</a></code></pre></div>
<pre><code>## Error in prepMultiAssay(objlist3, patient.data, dfmap) : 
##   ExperimentList does not have names, assign names</code></pre>
<p>Non-matching names may also be present in the <code>ExperimentList</code> elements and the “assay” column of the <code>sampleMap</code>. If names only differ by case and are identical and unique, names will be standardized to lower case and replaced.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(objlist3) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/chartr">toupper</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(objlist))</a>
<a class="sourceLine" id="cb65-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(objlist3)</a></code></pre></div>
<pre><code>## [1] "AFFY"        "METHYL 450K" "MIRNA"       "CNV GISTIC"</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(dfmap[, <span class="st">"assay"</span>])</a></code></pre></div>
<pre><code>## [1] Affy        Methyl 450k Mirna       CNV gistic 
## Levels: Affy Methyl 450k Mirna CNV gistic</code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb69-1" data-line-number="1"><span class="kw"><a href="../reference/prepMultiAssay.html">prepMultiAssay</a></span>(objlist3, patient.data, dfmap)<span class="op">$</span>experiments</a></code></pre></div>
<pre><code>## 
## Names in the ExperimentList do not match sampleMap assay
## standardizing will be attempted...</code></pre>
<pre><code>##  - names set to lowercase</code></pre>
<pre><code>## ExperimentList class object of length 4: 
##  [1] affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] methyl 450k: matrix with 2 rows and 5 columns 
##  [3] mirna: matrix with 4 rows and 3 columns 
##  [4] cnv gistic: RangedSummarizedExperiment with 5 rows and 4 columns</code></pre>
<p>When <code>colnames</code> in the <code>ExperimentList</code> cannot be matched back to the primary data (<code>colData</code>), these will be dropped and added to the drops element.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb73-1" data-line-number="1">exampleMap &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">sampleMap</a></span>(simpleMultiAssay2)</a>
<a class="sourceLine" id="cb73-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(doubleExp, colnames)</a></code></pre></div>
<pre><code>## $`methyl 2k`
## [1] "Jack"  "Jill"  "Bob"   "Bobby"
## 
## $`methyl 3k`
## [1] "Jack" "Jane" "Bob"</code></pre>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb75-1" data-line-number="1">exampleMap</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 3 columns
##       assay     primary     colname
##    &lt;factor&gt; &lt;character&gt; &lt;character&gt;
## 1 methyl 2k        Jack        Jack
## 2 methyl 2k        Jill        Jill
## 3 methyl 2k         Bob         Bob
## 4 methyl 3k        Jack        Jack
## 5 methyl 3k         Bob         Bob</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb77-1" data-line-number="1"><span class="kw"><a href="../reference/prepMultiAssay.html">prepMultiAssay</a></span>(doubleExp, patient.data, exampleMap)<span class="op">$</span>metadata<span class="op">$</span>drops</a></code></pre></div>
<pre><code>## 
## Not all colnames in the ExperimentList are found in the 
## sampleMap, dropping samples from ExperimentList...</code></pre>
<pre><code>## $`methyl 2k`
## [1] "Bobby"
## 
## $`methyl 3k`
## [1] "Jane"</code></pre>
<pre><code>## $`columns.methyl 2k`
## [1] "Bobby"
## 
## $`columns.methyl 3k`
## [1] "Jane"</code></pre>
<p>A similar operation is performed for checking “primary” <code>sampleMap</code> names and <code>colData</code> rownames. In this example, we add a row corresponding to “Joe” that does not have a match in the experimental data.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb81-1" data-line-number="1">exMap &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(dfmap,</a>
<a class="sourceLine" id="cb81-2" data-line-number="2">    <span class="kw"><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/DataFrame-class">DataFrame</a></span>(<span class="dt">assay =</span> <span class="st">"New methyl"</span>, <span class="dt">primary =</span> <span class="st">"Joe"</span>,</a>
<a class="sourceLine" id="cb81-3" data-line-number="3">        <span class="dt">colname =</span> <span class="st">"Joe"</span>))</a>
<a class="sourceLine" id="cb81-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/invisible">invisible</a></span>(<span class="kw"><a href="../reference/prepMultiAssay.html">prepMultiAssay</a></span>(objlist, patient.data, exMap))</a></code></pre></div>
<pre><code>## Warning in prepMultiAssay(objlist, patient.data, exMap): 
## Lengths of names in the ExperimentList and sampleMap
##  are not equal</code></pre>
<pre><code>## 
## Not all names in the primary column of the sampleMap
##  could be matched to the colData rownames; see $drops</code></pre>
<pre><code>## DataFrame with 1 row and 3 columns
##        assay     primary     colname
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt;
## 1 New methyl         Joe         Joe</code></pre>
<p>To create a <code>MultiAssayExperiment</code> from the results of the <code>prepMultiAssay</code> function, take each corresponding element from the resulting list and enter them as arguments to the <code>MultiAssayExperiment</code> constructor function.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" data-line-number="1">prepped &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prepMultiAssay.html">prepMultiAssay</a></span>(objlist, patient.data, exMap)</a></code></pre></div>
<pre><code>## Warning in prepMultiAssay(objlist, patient.data, exMap): 
## Lengths of names in the ExperimentList and sampleMap
##  are not equal</code></pre>
<pre><code>## 
## Not all names in the primary column of the sampleMap
##  could be matched to the colData rownames; see $drops</code></pre>
<pre><code>## DataFrame with 1 row and 3 columns
##        assay     primary     colname
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt;
## 1 New methyl         Joe         Joe</code></pre>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb89-1" data-line-number="1">preppedMulti &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment.html">MultiAssayExperiment</a></span>(prepped<span class="op">$</span>experiments, prepped<span class="op">$</span>colData,</a>
<a class="sourceLine" id="cb89-2" data-line-number="2">    prepped<span class="op">$</span>sampleMap, prepped<span class="op">$</span>metadata)</a>
<a class="sourceLine" id="cb89-3" data-line-number="3">preppedMulti</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 5 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>Alternatively, use the <code>do.call</code> function to easily create a <code>MultiAssayExperiment</code> from the output of <code>prepMultiAssay</code> function:</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb91-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>(MultiAssayExperiment, prepped)</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 5 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="helper-functions-to-create-bioconductor-classes-from-raw-data" class="section level2">
<h2 class="hasAnchor">
<a href="#helper-functions-to-create-bioconductor-classes-from-raw-data" class="anchor"></a>Helper functions to create <code>Bioconductor</code> classes from raw data</h2>
<p>Recent updates to the <code>GenomicRanges</code> and <code>SummarizedExperiment</code> packages allow the user to create standard <em>Bioconductor</em> classes from raw data. Raw data read in as either <code>data.frame</code> or <code>DataFrame</code> can be converted to <code>GRangesList</code> or <code>SummarizedExperiment</code> classes depending on the type of data.</p>
<p>The function to create a <code>GRangesList</code> from a <code>data.frame</code>, called <code>makeGRangesListFromDataFrame</code> can be found in the <code>GenomicRanges</code> package. <code>makeSummarizedExperimentFromDataFrame</code> is available in the <code>SummarizedExperiment</code> package. It is also possible to create a <code>RangedSummarizedExperiment</code> class object from raw data when ranged data is available.</p>
<p>A simple example can be obtained from the function documentation in <code>GenomicRanges</code>:</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb93-1" data-line-number="1">grlls &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">chr =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="st">"chr1"</span>, nrows), <span class="dt">start =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">11</span>, <span class="dv">15</span>),</a>
<a class="sourceLine" id="cb93-2" data-line-number="2">    <span class="dt">end =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">12</span>, <span class="dv">16</span>), <span class="dt">strand =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"+"</span>, <span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"*"</span>),</a>
<a class="sourceLine" id="cb93-3" data-line-number="3">    <span class="dt">score =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">5</span>), <span class="dt">specimen =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"a"</span>, <span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"b"</span>, <span class="st">"c"</span>),</a>
<a class="sourceLine" id="cb93-4" data-line-number="4">    <span class="dt">gene_symbols =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"GENE"</span>, letters[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq_len</a></span>(nrows)]))</a>
<a class="sourceLine" id="cb93-5" data-line-number="5"></a>
<a class="sourceLine" id="cb93-6" data-line-number="6">grldf &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(grlls, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb93-7" data-line-number="7"></a>
<a class="sourceLine" id="cb93-8" data-line-number="8">GRL &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicRanges/topics/makeGRangesListFromDataFrame">makeGRangesListFromDataFrame</a></span>(grldf, <span class="dt">split.field =</span> <span class="st">"specimen"</span>,</a>
<a class="sourceLine" id="cb93-9" data-line-number="9">    <span class="dt">names.field =</span> <span class="st">"gene_symbols"</span>)</a></code></pre></div>
<p>This can then be converted to a <code>RaggedExperiment</code> object for a rectangular representation that will conform more easily to the <code>MultiAssayExperiment</code> API requirements.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb94-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/RaggedExperiment/topics/RaggedExperiment-class">RaggedExperiment</a></span>(GRL)</a></code></pre></div>
<pre><code>## class: RaggedExperiment 
## dim: 5 3 
## assays(0):
## rownames(5): GENEa GENEb GENEc GENEd GENEe
## colnames(3): a b c
## colData names(0):</code></pre>
<p><em>Note</em>. See the <code>RaggedExperiment</code> vignette for more details.</p>
<p>In the <code>SummarizedExperiment</code> package:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb96-1" data-line-number="1">sels &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">chr =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="st">"chr2"</span>, nrows), <span class="dt">start =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">11</span>, <span class="dv">15</span>),</a>
<a class="sourceLine" id="cb96-2" data-line-number="2">    <span class="dt">end =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">12</span>, <span class="dv">16</span>), <span class="dt">strand =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"+"</span>, <span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"*"</span>),</a>
<a class="sourceLine" id="cb96-3" data-line-number="3">    <span class="dt">expr0 =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">3</span>, <span class="dv">7</span>), <span class="dt">expr1 =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">8</span>, <span class="dv">12</span>), <span class="dt">expr2 =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">12</span>, <span class="dv">16</span>))</a>
<a class="sourceLine" id="cb96-4" data-line-number="4">sedf &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(sels,</a>
<a class="sourceLine" id="cb96-5" data-line-number="5">    <span class="dt">row.names =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"GENE"</span>, letters[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rev">rev</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq_len</a></span>(nrows))]),</a>
<a class="sourceLine" id="cb96-6" data-line-number="6">    <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb96-7" data-line-number="7">sedf</a></code></pre></div>
<pre><code>##        chr start end strand expr0 expr1 expr2
## GENEe chr2    11  12      +     3     8    12
## GENEd chr2    12  13      -     4     9    13
## GENEc chr2    13  14      +     5    10    14
## GENEb chr2    14  15      *     6    11    15
## GENEa chr2    15  16      *     7    12    16</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb98-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/makeSummarizedExperimentFromDataFrame">makeSummarizedExperimentFromDataFrame</a></span>(sedf)</a></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 5 3 
## metadata(0):
## assays(1): ''
## rownames(5): GENEe GENEd GENEc GENEb GENEa
## rowData names(0):
## colnames(3): expr0 expr1 expr2
## colData names(0):</code></pre>
<p style="text-align: right;">
<a href="#header">back to top</a>
</p>
</div>
</div>
<div id="integrated-subsetting-across-experiments" class="section level1">
<h1 class="hasAnchor">
<a href="#integrated-subsetting-across-experiments" class="anchor"></a>Integrated subsetting across experiments</h1>
<p><code>MultiAssayExperiment</code> allows subsetting by rows, columns, and assays, rownames, and colnames, across all experiments simultaneously while guaranteeing continued matching of samples.</p>
<p>Subsetting can be done most compactly by the square bracket method, or more verbosely and potentially more flexibly by the <code>subsetBy*()</code> methods.</p>
<div id="subsetting-by-square-bracket" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-by-square-bracket" class="anchor"></a>Subsetting by square bracket <code>[</code>
</h2>
<p>The three positions within the bracket operator indicate rows, columns, and assays, respectively (pseudocode):</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb100-1" data-line-number="1">myMultiAssay[rows, columns, assays]</a></code></pre></div>
<p>For example, to select the gene “ENST00000355076”:</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb101-1" data-line-number="1">myMultiAssay[<span class="st">"ENST00000355076"</span>, , ]</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 7 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 2: 
##  [1] Affy: SummarizedExperiment with 1 rows and 4 columns 
##  [2] Methyl 450k: matrix with 1 rows and 5 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>The above operation works across all types of assays, whether ID-based (e.g. <code>matrix</code>, <code>ExpressionSet</code>, <code>SummarizedExperiment</code>) or range-based (e.g. <code>RangedSummarizedExperiment</code>). Note that when using the bracket method <code>[</code>, the drop argument is <em>TRUE</em> by default.</p>
<p>You can subset by rows, columns, and assays in a single bracket operation, and they will be performed in that order (rows, then columns, then assays). The following selects the <code>ENST00000355076</code> gene across all samples, then the first two samples of each assay, and finally the Affy and Methyl 450k assays:</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb104-1" data-line-number="1">myMultiAssay[<span class="st">"ENST00000355076"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Affy"</span>, <span class="st">"Methyl 450k"</span>)]</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 3 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 2: 
##  [1] Affy: SummarizedExperiment with 1 rows and 2 columns 
##  [2] Methyl 450k: matrix with 1 rows and 3 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="subsetting-by-character-integer-and-logical" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-by-character-integer-and-logical" class="anchor"></a>Subsetting by character, integer, and logical</h2>
<p>By columns - character, integer, and logical are all allowed, for example:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb107-1" data-line-number="1">myMultiAssay[, <span class="st">"Jack"</span>, ]</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 1 columns 
##  [2] Methyl 450k: matrix with 2 rows and 2 columns 
##  [3] Mirna: matrix with 4 rows and 1 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 1 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb109-1" data-line-number="1">myMultiAssay[, <span class="dv">1</span>, ]</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 1 columns 
##  [2] Methyl 450k: matrix with 2 rows and 2 columns 
##  [3] Mirna: matrix with 4 rows and 1 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 1 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb111-1" data-line-number="1">myMultiAssay[, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="ot">TRUE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>), ]</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 1 columns 
##  [2] Methyl 450k: matrix with 2 rows and 2 columns 
##  [3] Mirna: matrix with 4 rows and 1 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 1 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>By assay - character, integer, and logical are allowed:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb113-1" data-line-number="1">myMultiAssay[, , <span class="st">"Mirna"</span>]</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 13 sampleMap rows not in names(experiments)
##   removing 1 colData rownames not in sampleMap 'primary'</code></pre>
<pre><code>## A MultiAssayExperiment object of 1 listed
##  experiment with a user-defined name and respective class. 
##  Containing an ExperimentList class object of length 1: 
##  [1] Mirna: matrix with 4 rows and 3 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb116-1" data-line-number="1">myMultiAssay[, , <span class="dv">3</span>]</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 13 sampleMap rows not in names(experiments)
##   removing 1 colData rownames not in sampleMap 'primary'</code></pre>
<pre><code>## A MultiAssayExperiment object of 1 listed
##  experiment with a user-defined name and respective class. 
##  Containing an ExperimentList class object of length 1: 
##  [1] Mirna: matrix with 4 rows and 3 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb119-1" data-line-number="1">myMultiAssay[, , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="ot">FALSE</span>, <span class="ot">FALSE</span>, <span class="ot">TRUE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>)]</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 13 sampleMap rows not in names(experiments)
##   removing 1 colData rownames not in sampleMap 'primary'</code></pre>
<pre><code>## A MultiAssayExperiment object of 1 listed
##  experiment with a user-defined name and respective class. 
##  Containing an ExperimentList class object of length 1: 
##  [1] Mirna: matrix with 4 rows and 3 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="the-drop-argument" class="section level2">
<h2 class="hasAnchor">
<a href="#the-drop-argument" class="anchor"></a>the “drop” argument</h2>
<p>Specify <code>drop=FALSE</code> to keep assays with zero rows or zero columns, e.g.:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb122-1" data-line-number="1">myMultiAssay[<span class="st">"ENST00000355076"</span>, , , drop=<span class="ot">FALSE</span>]</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 1 rows and 4 columns 
##  [2] Methyl 450k: matrix with 1 rows and 5 columns 
##  [3] Mirna: matrix with 0 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 0 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>Using the default <code>drop=TRUE</code>, assays with no rows or no columns are removed:</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb124-1" data-line-number="1">myMultiAssay[<span class="st">"ENST00000355076"</span>, , , drop=<span class="ot">TRUE</span>]</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 7 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 2: 
##  [1] Affy: SummarizedExperiment with 1 rows and 4 columns 
##  [2] Methyl 450k: matrix with 1 rows and 5 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="more-on-subsetting-by-columns" class="section level2">
<h2 class="hasAnchor">
<a href="#more-on-subsetting-by-columns" class="anchor"></a>More on subsetting by columns</h2>
<p>Experimental samples are stored in the rows of <code>colData</code> but the columns of elements of <code>ExperimentList</code>, so when we refer to subsetting by columns, we are referring to columns of the experimental assays. Subsetting by samples / columns will be more obvious after recalling the <code>colData</code>:</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb127-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##              sex       age
##         &lt;factor&gt; &lt;integer&gt;
## Jack           M        38
## Jill           F        39
## Bob            M        40
## Barbara        F        41</code></pre>
<p>Subsetting by samples identifies the selected samples in rows of the colData DataFrame, then selects all columns of the <code>ExperimentList</code> corresponding to these rows. Here we use an integer to keep the first two rows of colData, and all experimental assays associated to those two primary samples:</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb129-1" data-line-number="1">myMultiAssay[, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 2 columns 
##  [2] Methyl 450k: matrix with 2 rows and 3 columns 
##  [3] Mirna: matrix with 4 rows and 1 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 2 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>Note that the above operation keeps different numbers of columns / samples from each assay, reflecting the reality that some samples may not have been assayed in all experiments, and may have replicates in some.</p>
<p>Columns can be subset using a logical vector. Here the dollar sign operator (<code>$</code>) accesses one of the columns in <code>colData</code>.</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb131-1" data-line-number="1">malesMultiAssay &lt;-<span class="st"> </span>myMultiAssay[, myMultiAssay<span class="op">$</span>sex <span class="op">==</span><span class="st"> "M"</span>]</a>
<a class="sourceLine" id="cb131-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(malesMultiAssay)</a></code></pre></div>
<pre><code>## DataFrame with 2 rows and 2 columns
##           sex       age
##      &lt;factor&gt; &lt;integer&gt;
## Jack        M        38
## Bob         M        40</code></pre>
<p>Finally, for special use cases you can exert detailed control of row or column subsetting, by using a <code>list</code> or <code>CharacterList</code> to subset. The following creates a <code>CharacterList</code> of the column names of each assay:</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" data-line-number="1">allsamples &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(myMultiAssay)</a>
<a class="sourceLine" id="cb133-2" data-line-number="2">allsamples</a></code></pre></div>
<pre><code>## CharacterList of length 4
## [["Affy"]] array1 array2 array3 array4
## [["Methyl 450k"]] methyl1 methyl2 methyl3 methyl4 methyl5
## [["Mirna"]] micro1 micro2 micro3
## [["CNV gistic"]] mysnparray1 mysnparray2 mysnparray3 mysnparray4</code></pre>
<p>Now let’s get rid of three Methyl 450k arrays, those in positions 3, 4, and 5:</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" data-line-number="1">allsamples[[<span class="st">"Methyl 450k"</span>]] &lt;-<span class="st"> </span>allsamples[[<span class="st">"Methyl 450k"</span>]][<span class="op">-</span><span class="dv">3</span><span class="op">:-</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb135-2" data-line-number="2">myMultiAssay[, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">as.list</a></span>(allsamples), ]</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 3 sampleMap rows with 'colname' not in colnames of experiments</code></pre>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 2 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb138-1" data-line-number="1"><span class="kw"><a href="../reference/subsetBy.html">subsetByColumn</a></span>(myMultiAssay,  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">as.list</a></span>(allsamples))  <span class="co">#equivalent</span></a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 3 sampleMap rows with 'colname' not in colnames of experiments</code></pre>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 2 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="subsetting-assays" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-assays" class="anchor"></a>Subsetting assays</h2>
<p>You can select certain assays / experiments using subset, by providing a character, logical, or integer vector. An example using character:</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb141-1" data-line-number="1">myMultiAssay[, , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Affy"</span>, <span class="st">"CNV gistic"</span>)]</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 8 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 2: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>You can subset assays also using logical or integer vectors:</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb144-1" data-line-number="1">is.cnv &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grepl</a></span>(<span class="st">"CNV"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(<span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">experiments</a></span>(myMultiAssay)))</a>
<a class="sourceLine" id="cb144-2" data-line-number="2">is.cnv</a></code></pre></div>
<pre><code>## [1] FALSE FALSE FALSE  TRUE</code></pre>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb146-1" data-line-number="1">myMultiAssay[, , is.cnv]  <span class="co">#logical subsetting</span></a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 12 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A MultiAssayExperiment object of 1 listed
##  experiment with a user-defined name and respective class. 
##  Containing an ExperimentList class object of length 1: 
##  [1] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb149-1" data-line-number="1">myMultiAssay[, , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(is.cnv)] <span class="co">#integer subsetting</span></a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 12 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A MultiAssayExperiment object of 1 listed
##  experiment with a user-defined name and respective class. 
##  Containing an ExperimentList class object of length 1: 
##  [1] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="subsetting-rows-features-by-ids-integers-or-logicals" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-rows-features-by-ids-integers-or-logicals" class="anchor"></a>Subsetting rows (features) by IDs, integers, or logicals</h2>
<p>Rows of the assays correspond to assay features or measurements, such as genes. Regardless of whether the assay is ID-based (e.g., <code>matrix</code>, <code>ExpressionSet</code>) or range-based (e.g., <code>RangedSummarizedExperiment</code>), they can be subset using any of the following:</p>
<ul>
<li><p>a <strong>character vector</strong> of IDs that will be matched to rownames in each assay</p></li>
<li><p>an <strong>integer vector</strong> that will select rows of this position from each assay. This probably doesn’t make sense unless every <code>ExperimentList</code> element represents the same measurements in the same order and will generate an error if any of the integer elements exceeds the number of rows in any <code>ExperimentList</code> element. The most likely use of integer subsetting would be as a <code>head</code> function, for example to look at the first 6 rows of each assay.</p></li>
<li><p>a <strong>logical vector</strong> that will be passed directly to the row subsetting operation for each assay.</p></li>
<li><p>a <strong>list</strong> or <strong>List</strong> with element names matching those in the <code>ExperimentList</code>. Each element of the subsetting list will be passed on exactly to subset rows of the corresponding element of the <code>ExperimentList</code>.</p></li>
</ul>
<p>Any <code>list</code> or <code>List</code> input allows for selective subsetting. The subsetting is applied only to the matching element names in the <code>ExperimentList</code>. For example, to only take the first two rows of the microRNA dataset, we use a named <code>list</code> to indicate what element we want to subset along with the <code>drop = FALSE</code> argument.</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb152-1" data-line-number="1">myMultiAssay[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">Mirna =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>), , ]</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 5 columns 
##  [3] Mirna: matrix with 2 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb154-1" data-line-number="1">## equivalently</a>
<a class="sourceLine" id="cb154-2" data-line-number="2"><span class="kw"><a href="../reference/subsetBy.html">subsetByRow</a></span>(myMultiAssay, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">Mirna =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>))</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 5 columns 
##  [3] Mirna: matrix with 2 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p>Again, these operations always return a <code>MultiAssayExperiment</code> class, unless <code>drop=TRUE</code> is passed to the <code>[</code> backet subset, with any <code>ExperimentList</code> element not containing the feature having zero rows.</p>
<p>For example, return a MultiAssayExperiment where <code>Affy</code> and <code>Methyl 450k</code> contain only “ENST0000035076”" row, and “Mirna” and “CNV gistic” have zero rows (<code>drop</code> argument is set to <code>FALSE</code> by default in <code>subsetBy*</code>):</p>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb156-1" data-line-number="1">featSub0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/subsetBy.html">subsetByRow</a></span>(myMultiAssay, <span class="st">"ENST00000355076"</span>)</a>
<a class="sourceLine" id="cb156-2" data-line-number="2">featSub1 &lt;-<span class="st"> </span>myMultiAssay[<span class="st">"ENST00000355076"</span>, , drop =<span class="st"> </span><span class="ot">FALSE</span>] <span class="co">#equivalent</span></a>
<a class="sourceLine" id="cb156-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/all.equal">all.equal</a></span>(featSub0, featSub1)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb158-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(featSub1)</a></code></pre></div>
<pre><code>## [1] "MultiAssayExperiment"
## attr(,"package")
## [1] "MultiAssayExperiment"</code></pre>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb160-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(<span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">experiments</a></span>(featSub1))</a></code></pre></div>
<pre><code>## [1] "ExperimentList"
## attr(,"package")
## [1] "MultiAssayExperiment"</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb162-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">experiments</a></span>(featSub1)</a></code></pre></div>
<pre><code>## ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 1 rows and 4 columns 
##  [2] Methyl 450k: matrix with 1 rows and 5 columns 
##  [3] Mirna: matrix with 0 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 0 rows and 4 columns</code></pre>
<p>In the following, <code>Affy</code> <code>SummarizedExperiment</code> keeps both rows but with their order reversed, and <code>Methyl 450k</code> keeps only its second row.</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb164-1" data-line-number="1">featSubsetted &lt;-</a>
<a class="sourceLine" id="cb164-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="../reference/subsetBy.html">subsetByRow</a></span>(myMultiAssay, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ENST00000355076"</span>, <span class="st">"ENST00000294241"</span>))</a>
<a class="sourceLine" id="cb164-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assay</a></span>(myMultiAssay, 1L)</a></code></pre></div>
<pre><code>##                 array1 array2 array3 array4
## ENST00000294241    101    103    105    107
## ENST00000355076    102    104    106    108</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb166-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assay</a></span>(featSubsetted, 1L)</a></code></pre></div>
<pre><code>##                 array1 array2 array3 array4
## ENST00000355076    102    104    106    108
## ENST00000294241    101    103    105    107</code></pre>
</div>
<div id="subsetting-rows-features-by-genomicranges" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-rows-features-by-genomicranges" class="anchor"></a>Subsetting rows (features) by <code>GenomicRanges</code>
</h2>
<p>For <code>MultiAssayExperiment</code> objects containing range-based objects (currently <code>RangedSummarizedExperiment</code>), these can be subset using a <code>GRanges</code> object, for example:</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb168-1" data-line-number="1">gr &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicRanges/topics/GRanges-class">GRanges</a></span>(<span class="dt">seqnames =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"chr1"</span>, <span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="dt">strand =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"+"</span>),</a>
<a class="sourceLine" id="cb168-2" data-line-number="2">              <span class="dt">ranges =</span> <span class="kw">IRanges</span>(<span class="dt">start =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">230602</span>, <span class="dv">443625</span>, <span class="dv">934533</span>),</a>
<a class="sourceLine" id="cb168-3" data-line-number="3">                               <span class="dt">end =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">330701</span>, <span class="dv">443724</span>, <span class="dv">934632</span>)))</a></code></pre></div>
<p>Now do the subsetting. The function doing the work here is <code><a href="https://www.rdocumentation.org/packages/IRanges/topics/findOverlaps-methods">IRanges::subsetByOverlaps</a></code> - see its arguments for flexible types of subsetting by range. The first three arguments here are for <code>subset</code>, the rest passed on to <code><a href="https://www.rdocumentation.org/packages/IRanges/topics/findOverlaps-methods">IRanges::subsetByOverlaps</a></code> through “…”:</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb169-1" data-line-number="1">subsetted &lt;-<span class="st"> </span><span class="kw"><a href="../reference/subsetBy.html">subsetByRow</a></span>(myMultiAssay, gr, <span class="dt">maxgap =</span> 2L, <span class="dt">type =</span> <span class="st">"within"</span>)</a>
<a class="sourceLine" id="cb169-2" data-line-number="2"><span class="kw"><a href="../reference/MultiAssayExperiment-methods.html">experiments</a></span>(subsetted)</a></code></pre></div>
<pre><code>## ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 0 rows and 4 columns 
##  [2] Methyl 450k: matrix with 0 rows and 5 columns 
##  [3] Mirna: matrix with 0 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 1 rows and 4 columns</code></pre>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb171-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class">rowRanges</a></span>(subsetted[[<span class="dv">4</span>]])</a></code></pre></div>
<pre><code>## GRanges object with 1 range and 1 metadata column:
##     seqnames        ranges strand |  feature_id
##        &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   b     chr1 240297-240396      - |     ID\\002
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Square bracket subsetting can still be used here, but passing on arguments to <code><a href="https://www.rdocumentation.org/packages/IRanges/topics/findOverlaps-methods">IRanges::subsetByOverlaps</a></code> through “…” is simpler using <code><a href="../reference/subsetBy.html">subsetByRow()</a></code>.</p>
</div>
<div id="subsetting-is-endomorphic" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-is-endomorphic" class="anchor"></a>Subsetting is endomorphic</h2>
<p><code>subsetByRow</code>, <code>subsetByColumn</code>, <code>subsetByAssay</code>, and square bracket subsetting are all “endomorphic” operations, in that they always return another <code>MultiAssayExperiment</code> object.</p>
</div>
<div id="double-bracket-subsetting-to-select-experiments" class="section level2">
<h2 class="hasAnchor">
<a href="#double-bracket-subsetting-to-select-experiments" class="anchor"></a>Double-bracket subsetting to select experiments</h2>
<p>A double-bracket subset operation refers to an experiment, and will return the object contained within an <code>ExperimentList</code> element. It is <strong>not</strong> endomorphic. For example, the first <code>ExperimentList</code> element is called “Affy” and contains a <code>SummarizedExperiment</code>:</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb173-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## [1] "Affy"        "Methyl 450k" "Mirna"       "CNV gistic"</code></pre>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb175-1" data-line-number="1">myMultiAssay[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 2 4 
## metadata(0):
## assays(1): ''
## rownames(2): ENST00000294241 ENST00000355076
## rowData names(0):
## colnames(4): array1 array2 array3 array4
## colData names(1): slope53</code></pre>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb177-1" data-line-number="1">myMultiAssay[[<span class="st">"Affy"</span>]]</a></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 2 4 
## metadata(0):
## assays(1): ''
## rownames(2): ENST00000294241 ENST00000355076
## rowData names(0):
## colnames(4): array1 array2 array3 array4
## colData names(1): slope53</code></pre>
<p style="text-align: right;">
<a href="#header">back to top</a>
</p>
</div>
</div>
<div id="helpers-for-data-clean-up-and-management" class="section level1">
<h1 class="hasAnchor">
<a href="#helpers-for-data-clean-up-and-management" class="anchor"></a>Helpers for data clean-up and management</h1>
<div id="complete-cases" class="section level2">
<h2 class="hasAnchor">
<a href="#complete-cases" class="anchor"></a><code>complete.cases</code>
</h2>
<p>The <code>complete.cases</code> function returns a logical vector of <code>colData</code> rows identifying which primary units have data for all experiments. Recall that <code>myMultiAssay</code> provides data for four individuals:</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb179-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##              sex       age
##         &lt;factor&gt; &lt;integer&gt;
## Jack           M        38
## Jill           F        39
## Bob            M        40
## Barbara        F        41</code></pre>
<p>Of these, only Jack has data for all 5 experiments:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb181-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/complete.cases">complete.cases</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## [1]  TRUE FALSE  TRUE  TRUE</code></pre>
<p>But all four have complete cases for Affy and Methyl 450k:</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb183-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/complete.cases">complete.cases</a></span>(myMultiAssay[, , <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</a></code></pre></div>
<pre><code>## [1] TRUE TRUE TRUE TRUE</code></pre>
<p>This output can be used to select individuals with complete data:</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb185-1" data-line-number="1">myMultiAssay[, <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/complete.cases">complete.cases</a></span>(myMultiAssay), ]</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 3 columns 
##  [2] Methyl 450k: matrix with 2 rows and 4 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 3 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="replicated" class="section level2">
<h2 class="hasAnchor">
<a href="#replicated" class="anchor"></a><code>replicated</code>
</h2>
<p>The <code>replicated</code> function identifies <code>primary</code> column values or biological units that have multiple observations per <code>assay</code>. It returns a <code>list</code> of <code>LogicalList</code>s that indicate what biological units have one or more replicate measurements. This output is used for merging replicates by default.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb187-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">replicated</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## $Affy
## LogicalList of length 4
## [["Jack"]] FALSE FALSE FALSE FALSE
## [["Jill"]] FALSE FALSE FALSE FALSE
## [["Barbara"]] FALSE FALSE FALSE FALSE
## [["Bob"]] FALSE FALSE FALSE FALSE
## 
## $`Methyl 450k`
## LogicalList of length 4
## [["Jack"]] TRUE TRUE FALSE FALSE FALSE
## [["Jill"]] FALSE FALSE FALSE FALSE FALSE
## [["Barbara"]] FALSE FALSE FALSE FALSE FALSE
## [["Bob"]] FALSE FALSE FALSE FALSE FALSE
## 
## $Mirna
## LogicalList of length 3
## [["Jack"]] FALSE FALSE FALSE
## [["Barbara"]] FALSE FALSE FALSE
## [["Bob"]] FALSE FALSE FALSE
## 
## $`CNV gistic`
## LogicalList of length 4
## [["Jack"]] FALSE FALSE FALSE FALSE
## [["Jill"]] FALSE FALSE FALSE FALSE
## [["Bob"]] FALSE FALSE FALSE FALSE
## [["Barbara"]] FALSE FALSE FALSE FALSE</code></pre>
</div>
<div id="intersectrows" class="section level2">
<h2 class="hasAnchor">
<a href="#intersectrows" class="anchor"></a><code>intersectRows</code>
</h2>
<p>The <code>intersectRows</code> function takes all common rownames across all experiments and returns a <code>MultiAssayExperiment</code> with those rows.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb189-1" data-line-number="1">(ensmblMatches &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">intersectRows</a></span>(myMultiAssay[, , <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]))</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 7 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## A MultiAssayExperiment object of 2 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 2: 
##  [1] Affy: SummarizedExperiment with 1 rows and 4 columns 
##  [2] Methyl 450k: matrix with 1 rows and 5 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb192-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(ensmblMatches)</a></code></pre></div>
<pre><code>## CharacterList of length 2
## [["Affy"]] ENST00000355076
## [["Methyl 450k"]] ENST00000355076</code></pre>
</div>
<div id="intersectcolumns" class="section level2">
<h2 class="hasAnchor">
<a href="#intersectcolumns" class="anchor"></a><code>intersectColumns</code>
</h2>
<p>A call to <code>intersectColumns</code> returns another <code>MultiAssayExperiment</code> where the columns of each element of the <code>ExperimentList</code> correspond exactly to the rows of <code>colData</code>. In many cases, this operation returns a 1-to-1 correspondence of samples to patients for each experiment assay unless replicates are present in the data.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb194-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">intersectColumns</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 3 columns 
##  [2] Methyl 450k: matrix with 2 rows and 4 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 3 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="mergereplicates" class="section level2">
<h2 class="hasAnchor">
<a href="#mergereplicates" class="anchor"></a><code>mergeReplicates</code>
</h2>
<p>The <code>mergeReplicates</code> function allows the user to specify a function (default: <code>mean</code>) for combining replicate columns in each assay element. This can be combined with <code>intersectColumns</code> to create a <code>MultiAssayExperiment</code> object with one measurement in each experiment per biological unit.</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb196-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">mergeReplicates</a></span>(<span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">intersectColumns</a></span>(myMultiAssay))</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 1 sampleMap rows with 'colname' not in colnames of experiments</code></pre>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 4: 
##  [1] Affy: SummarizedExperiment with 2 rows and 3 columns 
##  [2] Methyl 450k: matrix with 2 rows and 3 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 3 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
</div>
<div id="combine-c" class="section level2">
<h2 class="hasAnchor">
<a href="#combine-c" class="anchor"></a>combine <code>c</code>
</h2>
<p>The combine <code>c</code> function allows the user to append an experiment to the list of experiments already present in <code>MultiAssayExperiment</code>. In the case that additional observations on the same set of samples were performed, the <code>c</code> function can conveniently be referenced to an existing assay that contains the same ordering of sample measurements.</p>
<p>The <code>mapFrom</code> argument indicates what experiment has the exact same organization of samples that will be introduced by the new experiment dataset. If the number of columns in the new experiment do not match those in the reference experiment, an error will be thrown.</p>
<p>Here we introduce a toy dataset created on the fly:</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb199-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(myMultiAssay, <span class="dt">ExpScores =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>, <span class="dt">ncol =</span> <span class="dv">4</span>,</a>
<a class="sourceLine" id="cb199-2" data-line-number="2"><span class="dt">dim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"ENSMBL0001"</span>, <span class="st">"ENSMBL0002"</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"pt"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>))),</a>
<a class="sourceLine" id="cb199-3" data-line-number="3"><span class="dt">mapFrom =</span> 1L)</a></code></pre></div>
<pre><code>## Warning in .local(x, ...): Assuming column order in the data provided 
##  matches the order in 'mapFrom' experiment(s) colnames</code></pre>
<pre><code>## A MultiAssayExperiment object of 5 listed
##  experiments with user-defined names and respective classes. 
##  Containing an ExperimentList class object of length 5: 
##  [1] Affy: SummarizedExperiment with 2 rows and 4 columns 
##  [2] Methyl 450k: matrix with 2 rows and 5 columns 
##  [3] Mirna: matrix with 4 rows and 3 columns 
##  [4] CNV gistic: RangedSummarizedExperiment with 5 rows and 4 columns 
##  [5] ExpScores: matrix with 2 rows and 4 columns 
## Features: 
##  experiments() - obtain the ExperimentList instance 
##  colData() - the primary/phenotype DataFrame 
##  sampleMap() - the sample availability DataFrame 
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment 
##  *Format() - convert into a long or wide DataFrame 
##  assays() - convert ExperimentList to a SimpleList of matrices</code></pre>
<p><em>Note</em>: Alternatively, a <code>sampleMap</code> for the additional dataset can be provided.</p>
<p style="text-align: right;">
<a href="#header">back to top</a>
</p>
</div>
</div>
<div id="extractor-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#extractor-functions" class="anchor"></a>Extractor functions</h1>
<p>Extractor functions convert a <code>MultiAssayExperiment</code> into other forms that are convenient for analyzing. These would normally be called after any desired subsetting has been performed.</p>
<div id="getwithcoldata" class="section level2">
<h2 class="hasAnchor">
<a href="#getwithcoldata" class="anchor"></a><code>getWithColData</code>
</h2>
<p>Provides a single assay along with any associated ‘colData’ columns while keeping the assay class constant.</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb202-1" data-line-number="1">(affex &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">getWithColData</a></span>(myMultiAssay, 1L))</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 12 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## class: SummarizedExperiment 
## dim: 2 4 
## metadata(0):
## assays(1): ''
## rownames(2): ENST00000294241 ENST00000355076
## rowData names(0):
## colnames(4): Jack Jill Bob Barbara
## colData names(2): sex age</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb205-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(affex)</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##              sex       age
##         &lt;factor&gt; &lt;integer&gt;
## Jack           M        38
## Jill           F        39
## Bob            M        40
## Barbara        F        41</code></pre>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb207-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(affex)</a></code></pre></div>
<pre><code>## [1] "SummarizedExperiment"
## attr(,"package")
## [1] "SummarizedExperiment"</code></pre>
<p>It will error when the target data class does not support a <code>colData</code> replacement method, meaning that it typically works with <code>SummarizedExperiment</code> and <code>RaggedExperiment</code> assays and their extensions.</p>
</div>
<div id="longformat-wideformat" class="section level2">
<h2 class="hasAnchor">
<a href="#longformat-wideformat" class="anchor"></a><code>longFormat</code> &amp; <code>wideFormat</code>
</h2>
<p>Produces <em>long</em> (default) or <em>wide</em> <code>DataFrame</code> objects. The following produces a long <code>DataFrame</code> (the default) for the first two assays:</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb209-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">longFormat</a></span>(myMultiAssay[, , <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 7 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## DataFrame with 18 rows and 5 columns
##           assay     primary         rowname     colname     value
##     &lt;character&gt; &lt;character&gt;     &lt;character&gt; &lt;character&gt; &lt;integer&gt;
## 1          Affy        Jack ENST00000294241      array1       101
## 2          Affy        Jack ENST00000355076      array1       102
## 3          Affy        Jill ENST00000294241      array2       103
## 4          Affy        Jill ENST00000355076      array2       104
## 5          Affy     Barbara ENST00000294241      array3       105
## ...         ...         ...             ...         ...       ...
## 14  Methyl 450k        Jill ENST00000383706     methyl3         6
## 15  Methyl 450k     Barbara ENST00000355076     methyl4         7
## 16  Methyl 450k     Barbara ENST00000383706     methyl4         8
## 17  Methyl 450k         Bob ENST00000355076     methyl5         9
## 18  Methyl 450k         Bob ENST00000383706     methyl5        10</code></pre>
<p>This is especially useful for performing regression against patient or sample data from <code>colData</code> using the <code>pDataCols</code> argument:</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb212-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">longFormat</a></span>(myMultiAssay[, , <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">colDataCols=</span><span class="st">"age"</span>)</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 7 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## DataFrame with 18 rows and 6 columns
##           assay     primary         rowname     colname     value
##     &lt;character&gt; &lt;character&gt;     &lt;character&gt; &lt;character&gt; &lt;integer&gt;
## 1          Affy        Jack ENST00000294241      array1       101
## 2          Affy        Jack ENST00000355076      array1       102
## 3          Affy        Jill ENST00000294241      array2       103
## 4          Affy        Jill ENST00000355076      array2       104
## 5          Affy     Barbara ENST00000294241      array3       105
## ...         ...         ...             ...         ...       ...
## 14  Methyl 450k        Jill ENST00000383706     methyl3         6
## 15  Methyl 450k     Barbara ENST00000355076     methyl4         7
## 16  Methyl 450k     Barbara ENST00000383706     methyl4         8
## 17  Methyl 450k         Bob ENST00000355076     methyl5         9
## 18  Methyl 450k         Bob ENST00000383706     methyl5        10
##           age
##     &lt;integer&gt;
## 1          38
## 2          38
## 3          39
## 4          39
## 5          41
## ...       ...
## 14         39
## 15         41
## 16         41
## 17         40
## 18         40</code></pre>
<p>The “wide” format is useful for calculating correlations or performing regression against different genomic features. Wide format is in general not possible with replicate measurements, so we demonstrate on the cleaned <code>MultiAssayExperiment</code> for the first 5 columns:</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb215-1" data-line-number="1">maemerge &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">mergeReplicates</a></span>(<span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">intersectColumns</a></span>(myMultiAssay))</a></code></pre></div>
<pre><code>## harmonizing input:
##   removing 1 sampleMap rows with 'colname' not in colnames of experiments</code></pre>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb217-1" data-line-number="1"><span class="kw"><a href="../reference/MultiAssayExperiment-helpers.html">wideFormat</a></span>(maemerge, <span class="dt">colDataCols=</span><span class="st">"sex"</span>)[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 5 columns
##       primary      sex Affy_ENST00000294241 Affy_ENST00000355076
##   &lt;character&gt; &lt;factor&gt;            &lt;integer&gt;            &lt;integer&gt;
## 1        Jack        M                  101                  102
## 2         Bob        M                  107                  108
## 3     Barbara        F                  105                  106
##   Methyl.450k_ENST00000355076
##                     &lt;numeric&gt;
## 1                           2
## 2                           9
## 3                           7</code></pre>
</div>
<div id="assay-assays" class="section level2">
<h2 class="hasAnchor">
<a href="#assay-assays" class="anchor"></a><code>assay</code> / <code>assays</code>
</h2>
<p>The <code>assay</code> (singular) function takes a particular experiment and returns a matrix. By default, it will return the <em>first</em> experiment as a matrix.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb219-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assay</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>##                 array1 array2 array3 array4
## ENST00000294241    101    103    105    107
## ENST00000355076    102    104    106    108</code></pre>
<p>The <code>assays</code> (plural) function returns a <code>SimpleList</code> of data matrices from the <code>ExperimentList</code>:</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb221-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## List of length 4
## names(4): Affy Methyl 450k Mirna CNV gistic</code></pre>
</div>
</div>
<div id="the-cancer-genome-atlas-and-multiassayexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#the-cancer-genome-atlas-and-multiassayexperiment" class="anchor"></a>The Cancer Genome Atlas and MultiAssayExperiment</h1>
<p>Our most recent efforts include the release of the experiment data package, <code>curatedTCGAData</code>. This package will allow users to selectively download cancer datasets from The Cancer Genome Atlas (TCGA) and represent the data as <code>MultiAssayExperiment</code> objects. Please see the package vignette for more details.</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb223-1" data-line-number="1">BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"curatedTCGAData"</span>)</a></code></pre></div>
</div>
<div id="dimension-names-rownames-and-colnames" class="section level1">
<h1 class="hasAnchor">
<a href="#dimension-names-rownames-and-colnames" class="anchor"></a>Dimension names: <code>rownames</code> and <code>colnames</code>
</h1>
<p><code>rownames</code> and <code>colnames</code> return a <code>CharacterList</code> of row names and column names across all the assays. A <code>CharacterList</code> is an efficient alternative to <code>list</code> used when each element contains a character vector. It also provides a nice show method:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb224-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## CharacterList of length 4
## [["Affy"]] ENST00000294241 ENST00000355076
## [["Methyl 450k"]] ENST00000355076 ENST00000383706
## [["Mirna"]] hsa-miR-21 hsa-miR-191 hsa-miR-148a hsa-miR148b
## [["CNV gistic"]] a b c d e</code></pre>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb226-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(myMultiAssay)</a></code></pre></div>
<pre><code>## CharacterList of length 4
## [["Affy"]] array1 array2 array3 array4
## [["Methyl 450k"]] methyl1 methyl2 methyl3 methyl4 methyl5
## [["Mirna"]] micro1 micro2 micro3
## [["CNV gistic"]] mysnparray1 mysnparray2 mysnparray3 mysnparray4</code></pre>
<p style="text-align: right;">
<a href="#header">back to top</a>
</p>
</div>
<div id="requirements-for-support-of-additional-data-classes" class="section level1">
<h1 class="hasAnchor">
<a href="#requirements-for-support-of-additional-data-classes" class="anchor"></a>Requirements for support of additional data classes</h1>
<p>Any data classes in the <code>ExperimentList</code> object must support the following methods:</p>
<ul>
<li><code>dimnames</code></li>
<li><code>[</code></li>
<li><code><a href="https://www.rdocumentation.org/packages/base/topics/dim">dim()</a></code></li>
</ul>
<p>Here is what happens if one of the methods doesn’t:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb228-1" data-line-number="1">objlist2 &lt;-<span class="st"> </span>objlist</a>
<a class="sourceLine" id="cb228-2" data-line-number="2">objlist2[[<span class="dv">2</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">as.vector</a></span>(objlist2[[<span class="dv">2</span>]])</a>
<a class="sourceLine" id="cb228-3" data-line-number="3"></a>
<a class="sourceLine" id="cb228-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/try">try</a></span>(<span class="kw"><a href="../reference/MultiAssayExperiment.html">MultiAssayExperiment</a></span>(objlist2, patient.data, dfmap),</a>
<a class="sourceLine" id="cb228-5" data-line-number="5">    <span class="dt">outFile =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/showConnections">stdout</a></span>())</a></code></pre></div>
<pre><code>## Error in validObject(.Object) : invalid class "ExperimentList" object: 
##     Element [2] of class 'integer' does not have compatible method(s):
##     [</code></pre>
</div>
<div id="application-programming-interface-api" class="section level1">
<h1 class="hasAnchor">
<a href="#application-programming-interface-api" class="anchor"></a>Application Programming Interface (API)</h1>
<p>For more information on the formal API of <code>MultiAssayExperiment</code>, please see the <a href="https://github.com/waldronlab/MultiAssayExperiment/wiki/MultiAssayExperiment-API">API wiki</a> document on GitHub. An API package is available for download on GitHub via <code><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install("waldronlab/MultiAssayShiny")</a></code>. It provides visual exploration of available methods in <code>MultiAssayExperiment</code>.</p>
<p style="text-align: right;">
<a href="#header">back to top</a>
</p>
</div>
<div id="methods-for-multiassayexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#methods-for-multiassayexperiment" class="anchor"></a>Methods for MultiAssayExperiment</h1>
<p>The following methods are defined for <code>MultiAssayExperiment</code>:</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb230-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/methods">methods</a></span>(<span class="dt">class=</span><span class="st">"MultiAssayExperiment"</span>)</a></code></pre></div>
<pre><code>##  [1] [               [[              [[&lt;-            [&lt;-            
##  [5] $               $&lt;-             anyReplicated   assay          
##  [9] assays          c               coerce          colData        
## [13] colData&lt;-       complete.cases  dimnames        experiments    
## [17] experiments&lt;-   hasRowRanges    isEmpty         length         
## [21] mergeReplicates metadata        metadata&lt;-      names          
## [25] names&lt;-         replicated      sampleMap       sampleMap&lt;-    
## [29] show            subsetByAssay   subsetByColData subsetByColumn 
## [33] subsetByRow     updateObject   
## see '?methods' for accessing help and source code</code></pre>
</div>
<div id="citing-multiassayexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#citing-multiassayexperiment" class="anchor"></a>Citing MultiAssayExperiment</h1>
<p>We are excited to announce the official citation for MultiAssayExperiment in <em>Cancer Research</em>.</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb232-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/citation">citation</a></span>(<span class="st">"MultiAssayExperiment"</span>)</a></code></pre></div>
<pre><code>## 
## To cite MultiAssayExperiment in publications use:
## 
##   Marcel Ramos et al. Software For The Integration Of Multiomics
##   Experiments In Bioconductor. Cancer Research, 2017 November 1;
##   77(21); e39-42. DOI: 10.1158/0008-5472.CAN-17-0344
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     title = {Software For The Integration Of Multi-Omics Experiments In Bioconductor},
##     author = {Marcel Ramos and Lucas Schiffer and Angela Re and Rimsha Azhar and Azfar Basunia and Carmen Rodriguez Cabrera and Tiffany Chan and Philip Chapman and Sean Davis and David Gomez-Cabrero and Aedin C. Culhane and Benjamin Haibe-Kains and Kasper Hansen and Hanish Kodali and Marie Stephie Louis and Arvind Singh Mer and Markus Reister and Martin Morgan and Vincent Carey and Levi Waldron},
##     journal = {Cancer Research},
##     year = {2017},
##     volume = {77(21); e39-42},
##   }</code></pre>
</div>
<div id="sessioninfo" class="section level1">
<h1 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>sessionInfo()</h1>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb234-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()</a></code></pre></div>
<pre><code>## R Under development (unstable) (2019-02-12 r76095)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.5 LTS
## 
## Matrix products: default
## BLAS: /home/travis/R-bin/lib/R/lib/libRblas.so
## LAPACK: /home/travis/R-bin/lib/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] RaggedExperiment_1.7.4      MultiAssayExperiment_1.9.15
##  [3] SummarizedExperiment_1.13.0 DelayedArray_0.9.8         
##  [5] BiocParallel_1.17.13        matrixStats_0.54.0         
##  [7] Biobase_2.43.1              GenomicRanges_1.35.1       
##  [9] GenomeInfoDb_1.19.1         IRanges_2.17.4             
## [11] S4Vectors_0.21.10           BiocGenerics_0.29.1        
## [13] BiocStyle_2.11.0           
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.0             pillar_1.3.1           compiler_3.6.0        
##  [4] BiocManager_1.30.4     XVector_0.23.0         bitops_1.0-6          
##  [7] tools_3.6.0            zlibbioc_1.29.0        digest_0.6.18         
## [10] tibble_2.0.1           lattice_0.20-38        evaluate_0.13         
## [13] memoise_1.1.0          pkgconfig_2.0.2        rlang_0.3.1           
## [16] Matrix_1.2-15          commonmark_1.7         yaml_2.2.0            
## [19] pkgdown_1.3.0          xfun_0.4               GenomeInfoDbData_1.2.0
## [22] stringr_1.4.0          roxygen2_6.1.1         xml2_1.2.0            
## [25] knitr_1.21             desc_1.2.0             fs_1.2.6              
## [28] tidyselect_0.2.5       grid_3.6.0             rprojroot_1.3-2       
## [31] glue_1.3.0             R6_2.3.0               rmarkdown_1.11        
## [34] bookdown_0.9           purrr_0.3.0            tidyr_0.8.2           
## [37] magrittr_1.5           backports_1.1.3        htmltools_0.3.6       
## [40] MASS_7.3-51.1          assertthat_0.2.0       stringi_1.2.4         
## [43] RCurl_1.95-4.11        crayon_1.3.4</code></pre>
<p style="text-align: right;">
<a href="#header">back to top</a>
</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#installation">Installation</a></li>
      <li>
<a href="#a-brief-description">A Brief Description</a><ul class="nav nav-pills nav-stacked">
<li><a href="#choosing-the-appropriate-data-structure">Choosing the appropriate data structure</a></li>
      </ul>
</li>
      <li>
<a href="#overview-of-the-multiassayexperiment-class">Overview of the <code>MultiAssayExperiment</code> class</a><ul class="nav nav-pills nav-stacked">
<li><a href="#components-of-the-multiassayexperiment">Components of the <code>MultiAssayExperiment</code></a></li>
      </ul>
</li>
      <li>
<a href="#creating-a-multiassayexperiment-object-a-rich-example">Creating a <code>MultiAssayExperiment</code> object: a rich example</a><ul class="nav nav-pills nav-stacked">
<li><a href="#create-toy-datasets-demonstrating-all-supported-data-types">Create toy datasets demonstrating all supported data types</a></li>
      <li><a href="#samplemap-creation"><code>sampleMap</code> creation</a></li>
      <li><a href="#experimental-data-as-a-list">Experimental data as a <code><a href="https://www-rdocumentation-org/packages/base/topics/list">list()</a></code></a></li>
      <li><a href="#creation-of-the-multiassayexperiment-class-object">Creation of the <code>MultiAssayExperiment</code> class object</a></li>
      <li><a href="#helper-function-to-create-a-multiassayexperiment-object">Helper function to create a <code>MultiAssayExperiment</code> object</a></li>
      <li><a href="#helper-functions-to-create-bioconductor-classes-from-raw-data">Helper functions to create <code>Bioconductor</code> classes from raw data</a></li>
      </ul>
</li>
      <li>
<a href="#integrated-subsetting-across-experiments">Integrated subsetting across experiments</a><ul class="nav nav-pills nav-stacked">
<li><a href="#subsetting-by-square-bracket">Subsetting by square bracket <code>[</code></a></li>
      <li><a href="#subsetting-by-character-integer-and-logical">Subsetting by character, integer, and logical</a></li>
      <li><a href="#the-drop-argument">the “drop” argument</a></li>
      <li><a href="#more-on-subsetting-by-columns">More on subsetting by columns</a></li>
      <li><a href="#subsetting-assays">Subsetting assays</a></li>
      <li><a href="#subsetting-rows-features-by-ids-integers-or-logicals">Subsetting rows (features) by IDs, integers, or logicals</a></li>
      <li><a href="#subsetting-rows-features-by-genomicranges">Subsetting rows (features) by <code>GenomicRanges</code></a></li>
      <li><a href="#subsetting-is-endomorphic">Subsetting is endomorphic</a></li>
      <li><a href="#double-bracket-subsetting-to-select-experiments">Double-bracket subsetting to select experiments</a></li>
      </ul>
</li>
      <li>
<a href="#helpers-for-data-clean-up-and-management">Helpers for data clean-up and management</a><ul class="nav nav-pills nav-stacked">
<li><a href="#complete-cases"><code>complete.cases</code></a></li>
      <li><a href="#replicated"><code>replicated</code></a></li>
      <li><a href="#intersectrows"><code>intersectRows</code></a></li>
      <li><a href="#intersectcolumns"><code>intersectColumns</code></a></li>
      <li><a href="#mergereplicates"><code>mergeReplicates</code></a></li>
      <li><a href="#combine-c">combine <code>c</code></a></li>
      </ul>
</li>
      <li>
<a href="#extractor-functions">Extractor functions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getwithcoldata"><code>getWithColData</code></a></li>
      <li><a href="#longformat-wideformat"><code>longFormat</code> &amp; <code>wideFormat</code></a></li>
      <li><a href="#assay-assays"><code>assay</code> / <code>assays</code></a></li>
      </ul>
</li>
      <li><a href="#the-cancer-genome-atlas-and-multiassayexperiment">The Cancer Genome Atlas and MultiAssayExperiment</a></li>
      <li><a href="#dimension-names-rownames-and-colnames">Dimension names: <code>rownames</code> and <code>colnames</code></a></li>
      <li><a href="#requirements-for-support-of-additional-data-classes">Requirements for support of additional data classes</a></li>
      <li><a href="#application-programming-interface-api">Application Programming Interface (API)</a></li>
      <li><a href="#methods-for-multiassayexperiment">Methods for MultiAssayExperiment</a></li>
      <li><a href="#citing-multiassayexperiment">Citing MultiAssayExperiment</a></li>
      <li><a href="#sessioninfo">sessionInfo()</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Marcel Ramos, Levi Waldron.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
